"""Beginning of execution for `snakemake` command and start of workflow.
To change the output produced by the command, add required files to `input` under `rule all`.
The corresponding config file for the project can be written under `configfile`."""

configfile: "config/Mmul_8.yaml"
include: "rules/variant_calling.smk"
include: "rules/phasing.smk"
include: "rules/relations.smk"
wildcard_constraints:
    dataset="\w+"

rule all:
    """Generate target files. This rule runs automatically when Snakefile is run.
    Change the path under input to change the target files generated by snakemake.
    """
    #input: expand("{results}alignments_marked/{sample}.bam", results=config["results"], sample=SAMPLE_NAMES)
    #input: multiext(config["results"] + "ref/ref_genome.fna.gz", ".amb", ".ann", ".bwt", ".pac", ".sa")
    #input: config["results"] + "BQSR/GBS10235.BQSR.recal"
    #input: config["results"] + "joint-call/recalibrated_joint-call.relate"
    #input: config["results"] + "plink/split/GbS_2020.chr20.ped",
    #       config["results"] + "plink/split/GbS_2020.chr20.map"
    input: config["results"] + "haplotypes/split/GbS_2020.chr20.gen"
    #input: config["results"] + "haplotypes/GbS_2020.20.shapeit.haps"

