"""Beginning of execution for `snakemake` command and start of workflow.
To change the output produced by the command, add required files to `input` under `rule all`.
The corresponding config file for the project can be written under `configfile`."""

import sys
sys.path.insert(0, "./config")

from collections import defaultdict
import os
from pathlib import Path

# Create defaultdict of runs for samples.
# RUNS = defaultdict(set)
# for file in os.listdir(config["reads"]):
#     split = file.split(".")
#     RUNS[split[0]].add(split[1])


# Create list of runs
with open(config["runs"]) as f:
    SAMPLE_RUNS = f.read().splitlines()

#Create list of samples
if config["group_sample_runs_by_batch"]:
    tmp = 2
else:
    tmp = 1
SAMPLES = sorted(list(set([('_').join(run.split('/')[-1].split('_')[0:tmp]) for run in SAMPLE_RUNS])))

# Create list of chromosomes
chromosomes_file = config["resources"] + "ref_fna/chromosomes.list"
if Path(chromosomes_file).is_file():
    with open(chromosomes_file) as f:
        CHROMOSOMES = f.read().splitlines()
    print(CHROMOSOMES)
else:
    print("No chromosome file yet.")
    CHROMOSOMES = []
# CHROMOSOMES.remove("X")
# CHROMOSOMES.remove("Y")
# CHROMOSOMES.remove("MT")

# For command that use bash specific syntax, these have to activated. Although they may only work with the bio environment.
#shell.executable("/bin/bash")
shell.prefix("source ~/.bash_profile; ")

wildcard_constraints:
    dataset =r"\w+",
    filter_method = "hard_filtered|VQSR",
    #organism_id = "\w+",
    mode = "SNP|indel",
    name = r"[A-Za-z0-9_\.-]+",
    prefix = r"[A-Za-z0-9_/]+",
    path = r"[A-Za-z0-9_/\.-]+",
    #run = r"[A-Za-z0-9_-]+",
    sample = r"\w+",
    sample1 = r"\w+",
    sample2 = r"\w+",
    seq = "WGS|WES|GBS|AMP|merged",
    workspace=r"[A-Za-z0-9_\.]+",

include: "rules/directory_structure.smk"
include: "rules/indices.smk"
include: "rules/quality_control.smk"
include: "rules/align.smk"
include: "rules/variant_calling.smk"
#include: "rules/variant_recalibration.smk"
include: "rules/hard_filter.smk"
include: "rules/imputation.smk"
include: "rules/annotation.smk"
#include: "rules/gene_counts.smk"
include: "rules/phasing.smk"
#include: "rules/kinship.smk"
#include: "rules/relations.smk"
include: "rules/stats/TajimaD.smk"
include: "rules/stats/rare_variant_tests.smk"
include: "rules/scikit-allel.smk"
include: "rules/fst.smk"
include: "rules/pca.smk"
include: "rules/homozygosity.smk"
include: "rules/admixture.smk"
include: "rules/structural_variation.smk"
include: "rules/plot_alignments.smk"
include: "rules/concordance.smk"
include: "rules/coverage.smk"
include: "rules/one_off.smk"
include: "rules/misc.smk"

#ruleorder: concat_autosomes > bcf_to_vcfgz
#ruleorder: vep > bcf_to_vcfgz
#ruleorder: vcfgz_to_bcf > call_SVs
ruleorder: cut_adapters_with_i5_i7_genozipped > cut_adapters_with_i5_i7

rule all:
    """Generate target files. This rule runs automatically when Snakefile is run.
    Add or modify the paths under input to change the target files generated by snakemake.
    """
    input:
        # Input for generating README.md file
        #config["results"] + "README.md",
        config["resources"] + "ref_fna/chromosomes.list",
        config["target_files"],