"""Beginning of execution for `snakemake` command and start of workflow.
To change the output produced by the command, add required files to `input` under `rule all`.
The corresponding config file for the project can be written under `configfile`."""

import sys
sys.path.insert(0, "./config")

from collections import defaultdict
import os
from pathlib import Path

# Create defaultdict of runs for samples.
# RUNS = defaultdict(set)
# for file in os.listdir(config["reads"]):
#     split = file.split(".")
#     RUNS[split[0]].add(split[1])

# Create list of samples
with open(config["runs"]) as f:
    SAMPLE_RUNS = f.read().splitlines()

with open(config["samples"]) as f:
    SAMPLES = f.read().splitlines()

# with open(config["runs"]) as f:
#     SAMPLES = []
#     for row in f:
#         sample = row.split("_")[0]
#         if sample not in SAMPLES:
#             SAMPLES.append(sample)
# print(SAMPLES)

# Create list of chromosomes
if Path(config["results"] + "db/chromosomes.list").is_file():
    with open(config["results"] + "db/chromosomes.list") as f:
        CHROMOSOMES = f.read().splitlines()
    print(CHROMOSOMES)
else:
    print("No chromosome file yet.")
    CHROMOSOMES = []
# CHROMOSOMES.remove("X")
# CHROMOSOMES.remove("Y")
# CHROMOSOMES.remove("MT")

# For command that use bash specific syntax, these have to activated. Although they may only work with the bio environment.
#shell.executable("/bin/bash")
shell.prefix("source ~/.bash_profile; ")

wildcard_constraints:
    dataset ="\w+",
    filter_method = "hard_filtered|VQSR",
    #organism_id = "\w+",
    mode = "SNP|indel",
    name = "[A-Za-z0-9_\.-]+",
    prefix = "[A-Za-z0-9_/]+",
    path = "[A-Za-z0-9_/\.-]+",
    sample = "\w+",
    sample1 = "\w+",
    sample2 = "\w+",
    seq = "WGS|WES|GBS|AMP|merged",
    workspace="[A-Za-z0-9_\.]+",

include: "rules/directory_structure.smk"
include: "rules/indices.smk"
include: "rules/quality_control.smk"
include: "rules/variant_calling.smk"
#include: "rules/variant_recalibration.smk"
include: "rules/hard_filter.smk"
include: "rules/imputation.smk"
#include: "rules/gene_counts.smk"
include: "rules/phasing.smk"
#include: "rules/kinship.smk"
#include: "rules/relations.smk"
include: "rules/scikit-allel.smk"
include: "rules/fst.smk"
include: "rules/pca.smk"
include: "rules/homozygosity.smk"
include: "rules/admixture.smk"
include: "rules/structural_variation.smk"
include: "rules/concordance.smk"
include: "rules/coverage.smk"
include: "rules/one_off.smk"
include: "rules/misc.smk"

rule all:
    """Generate target files. This rule runs automatically when Snakefile is run.
    Add or modify the paths under input to change the target files generated by snakemake.
    """
    input:
        # Input for generating README.md file
        #config["results"] + "README.md",
        config["target_files"],
