"""Beginning of execution for `snakemake` command and start of workflow.
To change the output produced by the command, add required files to `input` under `rule all`.
The corresponding config file for the project can be written under `configfile`."""

configfile: "config/Mmul_10.yaml"
include: "rules/variant_calling.smk"
include: "rules/phasing.smk"
include: "rules/relations.smk"
wildcard_constraints:
    dataset="\w+"

rule all:
    """Generate target files. This rule runs automatically when Snakefile is run.
    Change the path under input to change the target files generated by snakemake.
    """
    # Input for generating README.md file
    input: config["results"] + "README.md"

    input: expand("{results}alignments_recalibrated/{sample}.bam", results=config["results"], sample=SAMPLE_NAMES)
   
rule create_readme:
    """Create README.md file in appropriate directory for easy reference about generated directories."""
    output: config["results"] + "README.md"
    shell: """echo \
    '''Information about where these files fall in the workflow can be seen under the `.smk` files under `workflow/rules`.

    A brief order and description is as follows:


    - ref  # Contatins reference genome information
    |
    - alignments  # .bam files after .fastq files aligned to reference
    |
    - alignments_rg  # @RG or "read group info" is added to .bam
    |
    - alignments_sorted  # Alignments sorted by position
    |
    - alignments_marked  # Duplicate sequences are marked
    |
    - BQSR  # Base recalibration tables for each sample
    |
    - alignments_recalibrated  # BQSR tables applied to sequences
    '''
    """

