"""Beginning of execution for `snakemake` command and start of workflow.
To change the output produced by the command, add required files to `input` under `rule all`.
The corresponding config file for the project can be written under `configfile`."""

configfile: "config/Mmul_10.yaml"
include: "rules/variant_calling.smk"
include: "rules/gene_counts.smk"
include: "rules/phasing.smk"
include: "rules/kinship.smk"
include: "rules/relations.smk"
include: "rules/coverage.smk"
wildcard_constraints:
    dataset="\w+"

rule all:
    """Generate target files. This rule runs automatically when Snakefile is run.
    Change the path under input to change the target files generated by snakemake.
    """
    
    input: # Input for generating README.md file
           config["results"] + "README.md",
           # User defined input goes below
           # These are two examples of how to structure desired files
           ## expand("{results}vcf/{sample}.g.vcf.gz", results=config["results"], sample=SAMPLE_NAMES)
           ## config["results"] + "db/cohort.sample_map"

   
rule create_readme:
    """Create README.md file in appropriate directory for easy reference about generated directories."""
    output: config["results"] + "README.md"
    shell: """echo '''
    Information about where these files fall in the workflow can be seen under the `.smk` files under `workflow/rules`.

    The directories for variant calling from `variant_calling.smk` are created in the following order:

    ├─ ref  # Contatins reference genome information
    |
    ├─ alignments  # .bam files after .fastq files aligned to reference
    |
    ├─ alignments_rg  # @RG or "read group info" is added to .bam
    |
    ├─ alignments_sorted  # Alignments sorted by position
    |
    ├─ alignments_marked  # Duplicate sequences are marked
    |
    ├─ BQSR  # Base recalibration tables for each sample
    |
    ├─ alignments_recalibrated  # BQSR tables applied to sequences
    |
    ├─ vcf  # Called variants as .g.vcf files
    |
    ├─ db  # GenomicsDB datastore consolidating .g.vcf files. Not human-readable
    |
    ├─ VQSR  # Variant recalibration
    |
    └─ joint_call  # db data converted to a joint vcf with VQSR applied


    The directories for phasing from `phasing.smk`:

    └─ haplotypes  # Phased 

    The directories for relations from `relations.smk` are as follows. `plink` is created first. The others do not follow an order.
    However, these require the joint vcf file from `joint_call`:

    ├─ plink  # Contains plink files
    |
    ├─ relatedness  # Pairwise estimates of relatedness between samples
    |
    ├─ admixture  # Find rates of admixture among samples
    |   |
    |   ├─ supervised  # Uses additional samples of known origin
    |   |
    |   └─ unsupervised  # No known origins
    |
    └─ aims  # Ancestry informative markers

    Directories for determining kinship using LASAR and SEEKIN:

    └─ kinship  # Allele frequencies and pairwise relatedness

    ''' > {output}
    """

