{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "\n",
    "#path = \"/master/abagwell/variant-analysis/results/rhesus/admixture/test/WGS_chr1_admix_by_birth_year.tsv\"\n",
    "path = \"/master/abagwell/variant-analysis/results/rhesus/admixture/test/chr1_admix_by_birth_year.tsv\"\n",
    "\n",
    "data = pl.read_csv(path, separator=\"\\t\", new_columns=[\"id\", \"Chinese\", \"Indian\", \"birth_date\"], has_header=False\n",
    ").filter(pl.col(\"id\").str.starts_with(\"m\").is_not()\n",
    ").with_columns(\n",
    "    (pl.col(\"Indian\") - 0.5).alias(\"scale\")\n",
    ").select(\n",
    "    \"id\",\n",
    "    \"Chinese\",\n",
    "    \"Indian\",\n",
    "    pl.col(\"birth_date\").str.to_date(\"%m-%d-%Y\").dt.year().alias(\"birth_year\"),\n",
    ").sort(\n",
    "    \"birth_year\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Overall admixture barplot. Second plot that displays number of births each year. These bars are selectable and modify the admixture plot.\n",
    "\n",
    "select_year = alt.selection_interval(encodings=['x'])\n",
    "selector = alt.selection_point(fields=[\"id\"])\n",
    "base = alt.Chart(data.to_arrow().to_pandas())\n",
    "\n",
    "heatmap = base.mark_bar().encode(\n",
    "    alt.X(\"Indian\", title=\"Fraction Indian\", scale=alt.Scale(domain=[0, 1])).bin(maxbins=20, extent=[0, 1]),\n",
    "    alt.Y(\"count(id)\", title=\"Counts\"),\n",
    "    #alt.Color('count()').legend(orient=\"right\", title=\"Paired Counts\"),\n",
    "    #tooltip=[\n",
    "    #    alt.Tooltip(\"count()\"),\n",
    "    #]\n",
    ").properties(\n",
    "    title=\"Indian vs Chinese Admixture in Rhesus Macaques\"\n",
    ").transform_filter(\n",
    "    select_year\n",
    ").add_params(\n",
    "    selector\n",
    ")\n",
    "\n",
    "births_per_year = base.mark_bar(width=7).encode(\n",
    "    alt.X(\"birth_year:O\", title='Birth Year'),\n",
    "    alt.Y(\"count()\", title='Count of Births'),\n",
    "    #color=alt.value(\"#dbe5b5\"),\n",
    "    #color=alt.value(\"#fc8d62\"),\n",
    "    color=alt.value(\"#66c2a5\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"count()\"),\n",
    "    ]\n",
    ").properties(\n",
    "    height = 70,\n",
    "    width = 500,\n",
    "    title = \"\"\n",
    ").add_selection(\n",
    "    select_year\n",
    ")\n",
    "\n",
    "heatmap & births_per_year"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ridgeline plot\n",
    "\n",
    "# Add duplicate rows to table, but with year adjusted. This allows for records to fall into multiple bins.\n",
    "tables = []\n",
    "for i in range(-2, 3):  # This range allows for a window that includes two below and two above the current year\n",
    "    table = data.with_columns(\n",
    "        pl.col(\"birth_year\").add(i) #apply(lambda y: y +1)\n",
    "    )\n",
    "    tables.append(table)\n",
    "\n",
    "data_extra_years = pl.concat(tables)\n",
    "\n",
    "step = 50\n",
    "overlap = 1.5\n",
    "\n",
    "def ridgeline(start):\n",
    "    return alt.Chart(data_extra_years.to_arrow().to_pandas(), height=step\n",
    "    ).transform_joinaggregate(\n",
    "        mean_Indian='mean(Indian)', groupby=['birth_year']\n",
    "    ).mark_area(\n",
    "        interpolate='monotone',\n",
    "        fillOpacity=0.8,\n",
    "        stroke='lightgray',\n",
    "        strokeWidth=0.5,\n",
    "    ).encode(\n",
    "        alt.X(\"Indian\").bin(maxbins=100).scale(domain=[start, 1], clamp=True).title(\"Avg. Indian\"),\n",
    "        alt.Y(\"count(id)\").axis(None).scale(range=[step, -step * overlap]).title(\"Count\"),\n",
    "        alt.Fill(\"mean_Indian:Q\").legend(None),#.scale(domain=[.35, .36], scheme='blueorange'),\n",
    "        alt.Row(\"birth_year:Q\").title(\"Birth Year\"),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(\"count(id)\", title=\"Samples in Bin\")\n",
    "        ]\n",
    "    ).properties(\n",
    "        title=\"Indian Ancestry by Birth Year (overlapping windows of 5 years)\",\n",
    "        bounds='flush',\n",
    "    ).configure_facet(\n",
    "        spacing=0,\n",
    "    ).configure_title(\n",
    "        anchor='middle',\n",
    "    )\n",
    "\n",
    "ridgeline(0) #+ ridgeline(0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Locus-specific Admixture\n",
    "import altair as alt\n",
    "from altair.expr import datum\n",
    "import polars as pl\n",
    "\n",
    "#path = \"/master/abagwell/variant-analysis/results/rhesus/admixture/RFMix.chr1.msp.tsv\"\n",
    "#path = \"/master/abagwell/variant-analysis/results/rhesus/admixture/RFMix.some.msp.tsv\"\n",
    "path = \"/master/abagwell/variant-analysis/results/rhesus/admixture/RFMix.all.msp.tsv\"\n",
    "\n",
    "data = pl.read_csv(path, separator=\"\\t\", skip_rows=1)\n",
    "\n",
    "# Unpack each genotype and sample into it's own record\n",
    "columns = data.columns\n",
    "\n",
    "unpacked = data.select(\n",
    "    pl.col(\"#chm\").alias(\"chm\"),\n",
    "    \"spos\",\n",
    "    \"epos\",\n",
    "    \"sgpos\",\n",
    "    \"egpos\",\n",
    "    \"n snps\",\n",
    "    pl.concat_list(pl.exclude(\"#chm\", \"spos\", \"epos\", \"sgpos\", \"egpos\", \"n snps\")).alias(\"genotype\"),\n",
    "    pl.Series([columns[6:] for record in range(len(data))]).alias(\"sample\")\n",
    ").explode(\"genotype\", \"sample\").with_columns(\n",
    "    pl.col(\"epos\").sub(pl.col(\"spos\")).alias(\"length\"),\n",
    "    pl.lit(0).alias(\"0\"),  # Create a column of all zeroes for starting position of y-axis in figure\n",
    ")#.sample(1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding intervals that are discrepant between WES and WGS. And then removing them.\n",
    "discordant = unpacked.with_columns(\n",
    "    pl.col(\"sample\").str.slice(0, 3).alias(\"seq_type\"),\n",
    "    pl.col(\"sample\").str.split(\"S\").arr.last().str.split(\".\").arr.first().alias(\"id\"),\n",
    ").groupby(\n",
    "    \"id\",\n",
    "    \"chm\",\n",
    "    \"spos\",\n",
    "    \"epos\",\n",
    "    #maintain_order=True\n",
    ").agg(  # Create list of WES/WGS for sample\n",
    "    \"seq_type\",\n",
    "    \"genotype\",\n",
    ").with_columns(  # Not required. Just to verify length of lists\n",
    "    pl.col(\"seq_type\").arr.lengths().alias(\"length\")\n",
    ").filter(  # Keeps only samples with both WES and WGS\n",
    "    pl.col(\"seq_type\").arr.lengths() == 4\n",
    ").filter(  # Keep only discordant regions between WES and WGS genotypes\n",
    "    (pl.col(\"genotype\").arr.get(0) != pl.col(\"genotype\").arr.get(2)) &\n",
    "    (pl.col(\"genotype\").arr.get(1) != pl.col(\"genotype\").arr.get(3)),\n",
    ").explode(  # Reverse the groupby\n",
    "    \"seq_type\",\n",
    "    \"genotype\"\n",
    ").with_columns(\n",
    "    pl.concat_str(pl.col(\"seq_type\"), pl.col(\"id\"), pl.lit(\".\"), pl.col(\"genotype\")).alias(\"sample\")\n",
    ")\n",
    "\n",
    "masked = unpacked.join(discordant, on=[\"chm\", \"spos\", \"epos\"], how=\"anti\")\n",
    "\n",
    "# Group bordering intervals that have the same genotype on the same chromosome for the same sample\n",
    "grouped = masked.sort(\"sample\").with_columns(\n",
    "    (\n",
    "        (pl.col(\"epos\") != pl.col(\"spos\").shift(-1)) |\n",
    "        (pl.struct(\"sample\", \"chm\", \"genotype\") != pl.struct(\"sample\", \"chm\", \"genotype\").shift(-1))\n",
    "    ).shift_and_fill(periods=1, fill_value=False).cumsum().alias(\"group\"),\n",
    ").groupby(\"group\").agg(\n",
    "    pl.col(\"chm\").first(),\n",
    "    pl.col(\"spos\").min().alias(\"spos\"),\n",
    "    pl.col(\"epos\").max().alias(\"epos\"),\n",
    "    pl.col(\"genotype\").first(),\n",
    "    pl.col(\"sample\").first(),\n",
    "    pl.col(\"length\").first(),\n",
    ")#.sort(\"group\")\n",
    "\n",
    "discordant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# discordant.filter(  # Remove WES that also have WGS\n",
    "#     pl.col(\"seq_type\") == \"WGS\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Long variable-width bar plot with dropdown for chromosome selection \n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "xlim = unpacked.select(\"epos\").max().item()\n",
    "\n",
    "# For chromosome selection\n",
    "chromosomes = unpacked.select(\"chm\").unique().to_series().to_list()\n",
    "input_dropdown = alt.binding_select(options=chromosomes, name='Chromosome')\n",
    "chr_selection = alt.param(value=1, bind=input_dropdown)\n",
    "\n",
    "alt.Chart(masked.to_arrow().to_pandas()).mark_bar(point=True).encode(\n",
    "    alt.X(\"spos\", title=\"Chromosomal Position (bp)\").scale(domain=[0, xlim], clamp=True),\n",
    "    alt.X2(\"epos\"),\n",
    "    alt.Y(\"0:Q\"),\n",
    "    alt.Y2(\"mean_genotype:Q\", title=\"Indian Admixture\"),\n",
    "    alt.Fill(\"mean_genotype:Q\").legend(None),\n",
    "    alt.Row(\"chm\", title=\"Chromosome\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"mean_genotype:Q\", title=\"Mean\"),\n",
    "        alt.Tooltip(\"spos\", title=\"Start\"),\n",
    "        alt.Tooltip(\"epos\", title=\"End\"),\n",
    "    ]\n",
    ").transform_aggregate(\n",
    "    mean_genotype='mean(genotype)',\n",
    "    groupby=[\"spos\", \"epos\", \"0\", \"chm\"],\n",
    ").properties(\n",
    "    title=\"Indian to Chinese Admixture Across Chromosome\",\n",
    "    width=20000\n",
    ").add_params(\n",
    "    chr_selection,\n",
    ").transform_filter(\n",
    "    (datum.chm == chr_selection)\n",
    ").save(\"admixture_variable-width_bar.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## For Manhattan plot\n",
    "chr_lengths = unpacked.select(\"chm\", \"epos\").groupby(\"chm\").max().sort(\"chm\")\n",
    "\n",
    "# Cumsum offsets of chromosomal end positions\n",
    "offsets = []\n",
    "offset = 0\n",
    "for chm, epos in chr_lengths.rows():\n",
    "    offsets.append((chm, offset))\n",
    "    offset += epos\n",
    "offsets = dict(offsets)\n",
    "\n",
    "# Add offsets\n",
    "concatenated = masked.with_columns(\n",
    "    pl.col(\"spos\").add(pl.col(\"chm\").map_dict(offsets)).alias(\"concat_spos\")\n",
    ")\n",
    "\n",
    "grouped_concat = grouped.with_columns(\n",
    "    pl.col(\"spos\").add(pl.col(\"chm\").map_dict(offsets)).alias(\"concat_spos\"),\n",
    "    pl.col(\"epos\").add(pl.col(\"chm\").map_dict(offsets)).alias(\"concat_epos\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manhattan plot\n",
    "alt.data_transformers.disable_max_rows()\n",
    "alt.Chart(concatenated.to_arrow().to_pandas()).mark_point(filled=True).encode(\n",
    "    alt.X(\"concat_spos\", title=\"Chromosomal Position (bp)\"), #.scale(domain=[0, xlim], clamp=True),\n",
    "    #alt.X2(\"epos\"),\n",
    "    alt.Y(\"mean_genotype:Q\", title=\"Indian Admixture\"),\n",
    "    #alt.Fill(\"mean_genotype:Q\").legend(None),\n",
    "    #alt.Column(\"chm\", align=\"each\", spacing=0, title=\"Chromosome\"),\n",
    "    color=alt.Color(\"chm:O\").scale(scheme='dark2').legend(None),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"mean_genotype:Q\", title=\"Mean\"),\n",
    "        alt.Tooltip(\"chm\", title=\"Chr\"),\n",
    "        alt.Tooltip(\"spos\", title=\"Start\"),\n",
    "        alt.Tooltip(\"epos\", title=\"End\"),\n",
    "    ]\n",
    ").transform_aggregate(\n",
    "    mean_genotype='mean(genotype)',\n",
    "    groupby=[\"spos\", \"epos\", \"concat_spos\", \"chm\"],\n",
    ").properties(\n",
    "    title=\"Indian to Chinese Admixture Across Chromosome\",\n",
    "    width=1000\n",
    ") #.save(\"admixture_Manhattan.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Horizontal bar graph. Each individual has two bars for each haplotype\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "chart = alt.Chart(grouped_concat.to_arrow().to_pandas()).mark_bar().encode(\n",
    "    alt.X(\"concat_spos\").title(\"Start\"),\n",
    "    alt.X2(\"concat_epos\").title(\"End\"),\n",
    "    alt.Y(\"sample\").title(\"Sample\"),\n",
    "    color = alt.Color(\"genotype:N\").scale(range=['#1f77b4', 'orange']),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"chm\", title=\"Chr\"),\n",
    "        alt.Tooltip(\"spos\", title=\"Start (bp)\"),\n",
    "        alt.Tooltip(\"epos\", title=\"End (bp)\"),\n",
    "        alt.Tooltip(\"length\", title=\"Length\"),\n",
    "        #alt.Tooltip(\"sgpos\", title=\"Start (cM)\"),\n",
    "        #alt.Tooltip(\"egpos\", title=\"Start (cM)\"),\n",
    "    ]\n",
    ").properties(\n",
    "    width=3000,\n",
    ")\n",
    "\n",
    "chromosome_boundaries = alt.Chart(pd.DataFrame({'x': list(offsets.values())[1:]})).mark_rule(color=\"red\", strokeWidth=8).encode(x='x')\n",
    "\n",
    "(chart + chromosome_boundaries) #.save(\"admixture_indiv_grouped.html\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
