{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Replace with path to merged BED file\n",
    "#file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/mosdepth/WGS_rhesus_1dup.merged.bed\"\n",
    "#file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/mosdepth/WGS_rhesus.merged.bed\"\n",
    "#file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/mosdepth/RPL.merged.bed\"\n",
    "file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/mosdepth/SNPRC_WES.merged.bed\"\n",
    "#file = \"/master/abagwell/variant-analysis/results/Caribbean/coverage/mosdepth/CPRC.merged.bed\"\n",
    "#file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/mosdepth/SNPRC_WGS_updated.merged.bed\"\n",
    "\n",
    "df = pd.read_table(file, names=[\"sample\", \"chrom\", \"start\", \"end\", \"avg_coverage\"])\n",
    "df = df[df[\"chrom\"].isin([str(i) for i in list(range(1,25)) + [\"X\", \"Y\", \"MT\"]])]  #Keep only numbered chromosomes\n",
    "df = df[df[\"end\"] - df[\"start\"] == 5_000_000]  # Remove the last interval (since it is not a full)\n",
    "\n",
    "not_normalized_df = df.copy()\n",
    "\n",
    "pos_coverage = df[[\"chrom\", \"start\", \"end\", \"avg_coverage\"]].groupby([\"chrom\", \"start\", \"end\"]).mean()\n",
    "pos_coverage_dict = pos_coverage.to_dict()[\"avg_coverage\"]\n",
    "#df.loc[df[\"sample\"] == \"31310\"].loc[df[\"chrom\"] != \"X\"].loc[df[\"chrom\"] != \"Y\"][\"avg_coverage\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Option 1: Seaborn implementation\n",
    "# This is a much simpler implementation, but is difficult to view when there are many samples,\n",
    "# is not modified for normalized coverage, and difficult to compare across chromosome.\n",
    "\n",
    "sns.set_theme()\n",
    "sns.relplot(data=df, hue=\"sample\", row=\"chrom\", x=\"start\", y=\"avg_coverage\", kind=\"line\", height=5, aspect=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Option 2: Bokeh implementation\n",
    "# More complex implementation but normalizes coverage, can be zoomed and expanded interactively, and displays sample labels when hovering.\n",
    "\n",
    "import itertools\n",
    "\n",
    "from bokeh.io import curdoc, output_file, output_notebook\n",
    "from bokeh.models import ColumnDataSource, WheelZoomTool, HoverTool, PanTool, ResetTool, Span\n",
    "#from bokeh.palettes import plasma as palette\n",
    "from bokeh.palettes import Category20 as palette\n",
    "palette = palette[20]\n",
    "from bokeh.plotting import figure, show\n",
    "import numpy as np\n",
    "\n",
    "samples = df[\"sample\"].unique()\n",
    "males = list(df.loc[df[\"chrom\"] == \"Y\"].loc[df[\"avg_coverage\"] < 5][\"sample\"])\n",
    "females = list(df.loc[df[\"chrom\"] == \"Y\"].loc[df[\"avg_coverage\"] >= 5][\"sample\"])\n",
    "chromosomes = [str(i) for i in list(range(1,21))] + [\"X\", \"Y\", \"MT\"]\n",
    "\n",
    "\n",
    "# Normalize samples\n",
    "for sample in samples:\n",
    "    #avg = df.loc[df[\"sample\"] == sample][\"avg_coverage\"].mean()\n",
    "    avg = df.loc[df[\"sample\"] == sample].loc[df[\"chrom\"] != \"X\"].loc[df[\"chrom\"] != \"Y\"][\"avg_coverage\"].mean()\n",
    "    #print(df.loc[df[\"sample\"] == sample][\"avg_coverage\"])\n",
    "    #df[\"normalized_coverage\"] = df.loc[:, \"avg_coverage\"]\n",
    "    df.loc[df[\"sample\"] == sample, [\"avg_coverage\"]] /= avg\n",
    "\n",
    "    # min = df[\"avg_coverage\"].min()\n",
    "    # range = df[\"avg_coverage\"].max() - min\n",
    "    # df.loc[df[\"sample\"] == sample, [\"avg_coverage\"]] -= min\n",
    "    # df.loc[df[\"sample\"] == sample, [\"avg_coverage\"]] /= range\n",
    "\n",
    "    \n",
    "    #df.loc[df[\"sample\"] == sample, [\"avg_coverage\"]] = np.log2(df.loc[df[\"sample\"] == sample, [\"avg_coverage\"]]) # Normalize from 0 to 1\n",
    "\n",
    "# Normalize by \n",
    "\n",
    "# # Normalize by position\n",
    "# for index, row in df.iterrows():\n",
    "#     pos_cov = pos_coverage_dict[(row[\"chrom\"], row[\"start\"], row[\"end\"])]\n",
    "#     row[\"avg_coverage\"] = pos_cov /row[\"avg_coverage\"]\n",
    "\n",
    "\n",
    "\n",
    "# Concat chromosomes\n",
    "end = 0\n",
    "chromosome_endpoints = []\n",
    "for chrom in chromosomes:\n",
    "    df.loc[df[\"chrom\"] == chrom, [\"start\"]] += end\n",
    "    end = df.loc[df[\"chrom\"] == chrom][\"start\"].max()\n",
    "    if chrom != \"MT\":  # Remove last endpoint\n",
    "        chromosome_endpoints.append(end)\n",
    "\n",
    "#df.loc[df[\"sample\"] == sample].loc[df[\"chrom\"] != \"X\"].loc[df[\"chrom\"] != \"Y\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov_descriptor = \"Normalized Coverage\"\n",
    "\n",
    "tooltips = [\n",
    "    (\"sample\", \"@sample\"),\n",
    "    (\"Window\", \"@start - @end\"),\n",
    "    (\"Chrom\", \"@chrom\"),\n",
    "    (cov_descriptor, \"@avg_coverage\")\n",
    "]\n",
    "\n",
    "# Create figure\n",
    "p = figure(title=f\"{cov_descriptor} of RPL Rhesus Macaques\", x_axis_label='Genome Postition', y_axis_label=cov_descriptor,\n",
    "    tools=[WheelZoomTool(), PanTool(), HoverTool(line_policy=\"nearest\"), ResetTool()], tooltips=tooltips)\n",
    "\n",
    "# Add lines\n",
    "colors = itertools.cycle(palette)\n",
    "for sample in samples:\n",
    "    #if sample not in [\"WGS17534\", \"WGS33938\", \"WGS33963\"]:\n",
    "    color = next(colors)\n",
    "    # if sample in males:\n",
    "    #     color = \"teal\"\n",
    "    # else:\n",
    "    #     color = \"orange\"\n",
    "    subset = df[df[\"sample\"] == sample]\n",
    "    source = ColumnDataSource(subset)\n",
    "    p.line(x=\"start\", y=\"avg_coverage\", legend_label=sample, color=color, alpha=0.3, hover_line_color=\"black\", hover_alpha=1, line_width=1, source=source)\n",
    "\n",
    "# Modify properties\n",
    "p.legend.label_text_font_size = \"8px\"\n",
    "p.legend.title = \"Samples\"\n",
    "p.legend.visible = False  # Hides legend\n",
    "p.sizing_mode = \"stretch_width\"\n",
    "p.height = 700\n",
    "\n",
    "# Add spans to separate chromosomes\n",
    "for endpoint in chromosome_endpoints:\n",
    "    p.add_layout(Span(location=endpoint, dimension=\"height\", line_width=3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output figure\n",
    "curdoc().theme = \"light_minimal\"\n",
    "output_notebook()\n",
    "#output_file(\"WGS_coverage_of_rhesus_macaques_color_outliers_included.html\")\n",
    "output_file(f\"/master/abagwell/figures/coverage/{cov_descriptor}_of_RPL_rhesus_macaques_color_outliers_included.html\")\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output figure\n",
    "curdoc().theme = \"light_minimal\"\n",
    "#output_notebook()\n",
    "output_file(\"WGS_coverage_of_rhesus_macaques_outliers_removed.html\")\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 3: Altair implementation (best)\n",
    "import altair as alt\n",
    "import polars as pl\n",
    "\n",
    "pl_df = pl.from_pandas(df).with_columns(\n",
    "    pl.col(\"end\").sub(5_000_000).alias(\"original_start\"),\n",
    ").to_arrow().to_pandas()\n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "base = alt.Chart(pl_df)\n",
    "cov_descriptor = \"Normalized Coverage\"\n",
    "\n",
    "#selector = alt.selection_point(fields=[\"sample\"])\n",
    "selector = alt.selection_interval()\n",
    "#color=alt.condition(selector, \"sample\", alt.value(\"lightgray\"))\n",
    "\n",
    "chr_lengths = df[[\"chrom\", \"end\"]].groupby(\"chrom\", sort=False).max(\"end\")\n",
    "#chr_lengths = largest_seq.select(\"chm\", \"epos\").groupby(\"chm\").max().sort(\"chm\")\n",
    "\n",
    "# Cumsum offsets of chromosomal end positions\n",
    "# offsets = []\n",
    "# offset = 0\n",
    "# for chm, epos in chr_lengths.iterrows():\n",
    "#     epos = epos[0]\n",
    "#     offsets.append((chm, offset))\n",
    "#     offset += epos\n",
    "# offsets = dict(offsets)\n",
    "offsets = [pos for pos in chromosome_endpoints]\n",
    "\n",
    "lineplot = base.mark_line().encode(\n",
    "    alt.X(\"start\", title=\"Genome Position\", axis=alt.Axis(grid=False)),\n",
    "    alt.Y('mean(avg_coverage)', title=cov_descriptor),  #(\"count()\"),   #, scale=alt.Scale(domain=[0,35])),\n",
    "    color=alt.condition(selector, alt.Color(\"sample:N\").legend(None).scale(scheme=\"set2\"), alt.value(\"lightgray\")),\n",
    "    opacity = alt.condition(selector, alt.value(1.0), alt.value(0.0)),\n",
    "    #color=alt.Color(\"sample:N\").legend(None).scale(scheme=\"set2\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"sample\", title=\"Sample\"),\n",
    "        #alt.Tooltip(\"end\" - 5_000_000, title=\"Start\"),\n",
    "        alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "        alt.Tooltip(\"original_start\", title=\"Start\"),\n",
    "        alt.Tooltip(\"end\", title=\"End\"),\n",
    "        alt.Tooltip(\"avg_coverage\", title=cov_descriptor)\n",
    "    ]\n",
    ").properties(\n",
    "    #width = 1200, # Original\n",
    "    width = 4000,\n",
    "    #title = f\"{cov_descriptor} of SNPRC Rhesus Macaques\",\n",
    "    title = f\"{cov_descriptor} of WGS Rhesus Macaques\",\n",
    ").interactive()\n",
    "\n",
    "chromosome_boundaries = alt.Chart(pd.DataFrame({'x': offsets})).mark_rule(color=\"gray\", strokeWidth=1).encode(x='x')\n",
    "\n",
    "selectable_samples = alt.Chart(not_normalized_df).mark_bar(width=7).encode(\n",
    "    alt.X(\"sample\", title='Samples'),#.sort(\"-y\"),\n",
    "    alt.Y(\"mean(avg_coverage)\", title='Coverage'),\n",
    "    color=alt.value(\"#93C572\"),\n",
    "    # tooltip=[\n",
    "    #     alt.Tooltip(\"count()\"),\n",
    "    # ]\n",
    ").properties(\n",
    "    height = 50,  # WGS_rhesus\n",
    "    #width = 700,  # RPL\n",
    "    #width = 1000,  # CPRC\n",
    "    width = 4000, # WES\n",
    "    #width = 1400, # mew WGS\n",
    "    title = f\"Samples by Coverage (selectable)\",\n",
    ").add_params(\n",
    "    selector\n",
    ")\n",
    "\n",
    "chromosome_boundaries + lineplot & selectable_samples\n",
    "(chromosome_boundaries + lineplot & selectable_samples).save(\"/master/abagwell/figures/normalized_coverage_SNPRC_WES_altair2.html\")\n",
    "#(chromosome_boundaries + lineplot & selectable_samples).save(\"/master/abagwell/figures/normalized_coverage_SNPRC_WGS_altair.html\")\n",
    "#(chromosome_boundaries + lineplot & selectable_samples).save(\"/master/abagwell/figures/normalized_coverage_CPRC_altair.html\")\n",
    "#(chromosome_boundaries + lineplot & selectable_samples).save(\"/master/abagwell/figures/normalized_coverage_RPL_altair.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Altair plot with chromosomes split\n",
    "# Under development\n",
    "\n",
    "plot_list = []\n",
    "for idx, chrom in enumerate(chromosomes):\n",
    "    lineplot = alt.Chart(pl_df).mark_line().encode(\n",
    "    alt.X(\"start\", title=\"Genome Position\", axis=alt.Axis(grid=False)),\n",
    "    alt.Y('mean(avg_coverage)', title=cov_descriptor),  #(\"count()\"),   #, scale=alt.Scale(domain=[0,35])),\n",
    "    color=alt.condition(selector, alt.Color(\"sample:N\").legend(None).scale(scheme=\"set2\"), alt.value(\"lightgray\")),\n",
    "    opacity = alt.condition(selector, alt.value(1.0), alt.value(0.0)),\n",
    "    #color=alt.Color(\"sample:N\").legend(None).scale(scheme=\"set2\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"sample\", title=\"Sample\"),\n",
    "        #alt.Tooltip(\"end\" - 5_000_000, title=\"Start\"),\n",
    "        alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "        alt.Tooltip(\"original_start\", title=\"Start\"),\n",
    "        alt.Tooltip(\"end\", title=\"End\"),\n",
    "        alt.Tooltip(\"avg_coverage\", title=cov_descriptor)\n",
    "    ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        #title = f\"{cov_descriptor} of SNPRC Rhesus Macaques\",\n",
    "        title = f\"{cov_descriptor} of WES Rhesus Macaques\",\n",
    "    ).interactive()\n",
    "\n",
    "selectable_samples = alt.Chart(not_normalized_df).mark_bar(width=7).encode(\n",
    "    alt.X(\"sample\", title='Samples').sort(\"-y\"),\n",
    "    alt.Y(\"mean(avg_coverage)\", title='Coverage'),\n",
    "    color=alt.value(\"#93C572\"),\n",
    "    # tooltip=[\n",
    "    #     alt.Tooltip(\"count()\"),\n",
    "    # ]\n",
    ").properties(\n",
    "    height = 50,  # WGS_rhesus\n",
    "    width = 700,  # RPL\n",
    "    #width = 1000,  # CPRC\n",
    "    title = f\"Samples by Coverage (selectable)\"\n",
    ").add_params(\n",
    "    selector\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boxplot of chrX\n",
    "# Also removing specific samples that had different pattern of coverage as well as chrY\n",
    "\n",
    "# Place sample to remove here\n",
    "deviant_samples = []\n",
    "\n",
    "pl_df = pl.from_pandas(df).groupby(\n",
    "    pl.col(\"sample\"),\n",
    "    pl.col(\"chrom\"),\n",
    ").agg(\n",
    "    pl.col(\"avg_coverage\").mean()\n",
    ").sort(\"chrom\").filter(\n",
    "    ~pl.col(\"sample\").is_in(deviant_samples)\n",
    ")\n",
    "# .filter(\n",
    "#     pl.col(\"chrom\") != \"Y\"\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_order = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"X\", \"Y\"]\n",
    "alt.Chart(pl_df.to_arrow().to_pandas()).mark_boxplot().encode(\n",
    "    alt.X(\"chrom\", title=\"Chromosome\", sort=chrom_order),\n",
    "    alt.Y(\"avg_coverage\", title=\"Normalized Coverage\").scale(zero=False),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"sample\", title=\"Sample\"),\n",
    "    ],\n",
    "    color=alt.Color(\"chrom\", sort=chrom_order, legend=None).scale(scheme=\"dark2\") #.scale(range=['#1f77b4', 'orange']),\n",
    ").properties(\n",
    "    #width=40,\n",
    "    #title=[\"chrX\"],\n",
    "    title=\"Normalized Coverage of Rhesus RPL\",\n",
    ")#.save(\"/master/abagwell/figures/WES_boxplot.svg\")\n",
    "#.save(\"/master/abagwell/figures/WES_boxplot_excluding5_and_chrY.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('graph')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "de4a875ba73fa07c7c0f7a1dd2e6901433a766d080dc9d908c0725e19addd1ed"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
