{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge ROH files and filter\n",
    "\n",
    "import altair as alt\n",
    "import polars as pl\n",
    "\n",
    "# Read colony info\n",
    "colonies_file = \"/master/abagwell/variant-analysis/resources/rhesus/pop/colonies.tsv\"\n",
    "colonies = pl.read_csv(colonies_file, separator=\"\\t\", infer_schema_length=None,schema_overrides={\n",
    "    \"Dam\": pl.String,\n",
    "    \"Sire\": pl.String,\n",
    "    \"Id\": pl.String\n",
    "})\n",
    "\n",
    "chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "chromosome_lengths = [223_616_942, 196_197_964, 185_288_947, 169_963_040, 187_317_192, 179_085_566, 169_868_564, 145_679_320, 134_124_166, 99_517_758, 133_066_086, 130_043_856, 108_737_130, 128_056_306, 113_283_604, 79_627_064, 95_433_459, 74_474_043, 58_315_233, 77_137_495]  # For Mmul_10\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## For when using `bcftools roh`\n",
    "# Changes column names to match scikit-allel\n",
    "\n",
    "roh_dfs = []\n",
    "for chrom in chromosomes:\n",
    "    roh_file = f\"/master/abagwell/variant-analysis/results/rhesus/roh/bcftools/U42_WES.common_between_founding_cohorts.chr{chrom}.RG.roh\"\n",
    "    roh_df = pl.read_csv(roh_file, comment_prefix=\"#\", has_header=False, separator=\"\\t\",\n",
    "        schema_overrides=[pl.String, pl.String, pl.String, pl.Int32, pl.Int32, pl.Int64, pl.Int32, pl.Float32],\n",
    "        new_columns=[\"RG\", \"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Number of markers\", \"Quality (average fwd-bwd phred score)\"]\n",
    "    ).filter(\n",
    "        # Keep only large ROHs\n",
    "        pl.col(\"length\") > 1_000_000\n",
    "    ).with_columns(\n",
    "    ).join(colonies, how=\"left\", left_on=\"sample\", right_on=\"Id\").drop(\"tag\")\n",
    "    \n",
    "    roh_dfs.append(roh_df)\n",
    "\n",
    "roh_df = pl.concat(roh_dfs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_trio_roh(df):\n",
    "    \"\"\"Plot the ROHs across all chromosomes of one trio.\"\"\"\n",
    "\n",
    "    # Create dummy records. That is, an ROH of length 0 for every combination of sample and chromosome so that they all appear on the plot\n",
    "    chromosomes = roh_df.select(\"chrom\").unique()\n",
    "    sequenced_animals = df.select(\"sample\", \"Colony\").unique()\n",
    "    dummy_records = sequenced_animals.join(chromosomes, how=\"cross\").with_columns(\n",
    "        start = pl.lit(0).cast(pl.Int32),\n",
    "        stop = pl.lit(0).cast(pl.Int32),\n",
    "        length = pl.lit(0).cast(pl.Int64),\n",
    "    ).select(  # Reorder\n",
    "        \"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Colony\"\n",
    "    )\n",
    "\n",
    "    # TODO: Ideally don't have this redundancy\n",
    "    chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "\n",
    "    # Concatenate ROHs with dummy records\n",
    "    df = pl.concat([df, dummy_records]).sort(\"sample\")  # TODO: Improve sorting for when needs to be numeric or animal ID has \"X\" in it\n",
    "\n",
    "    columns = []\n",
    "    for idx, chrom in enumerate(chromosomes):\n",
    "        #column_plots = []\n",
    "        # Only keep y-axis labels on leftmost plot\n",
    "        if idx == 0:\n",
    "            axis = alt.Axis()\n",
    "        else:\n",
    "            axis = None\n",
    "        # # Info from `bcftools roh`\n",
    "        chrom_df = df.filter(\n",
    "            # Keep only large ROHs\n",
    "            #pl.col(\"length\") > 2_000_000\n",
    "            pl.col(\"chrom\") == chrom\n",
    "        )\n",
    "\n",
    "        roh_plot = alt.Chart(chrom_df).mark_bar().encode(\n",
    "        alt.X(\"mb_start_pos:Q\", title=[\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),\n",
    "        alt.X2(\"mb_stop_pos:Q\"),\n",
    "        alt.Y(\"sample:O\", axis=axis, title=\"Sample\", sort=alt.EncodingSortField(field='animal', order='ascending')), # sort=alt.EncodingSortField(field='date', order='ascending')\n",
    "        color = alt.Color(\"Colony:N\"),  # .scale(range=['#1f77b4', 'orange', '#9F2B68'])\n",
    "        #order=alt.Order('animal:O', sort='ascending'),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "            alt.Tooltip(\"start\", title=\"Start (bp)\"),\n",
    "            alt.Tooltip(\"stop\", title=\"End (bp)\"),\n",
    "            alt.Tooltip(\"length\", title=\"Length\"),\n",
    "        ]\n",
    "        ).properties(\n",
    "            width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "            height=alt.Step(10),\n",
    "            #width=300, # For when only displaying one chromosome\n",
    "            #title=[\"RoH\", f\"chr{chrom}\"]\n",
    "        ).transform_calculate(\n",
    "            mb_start_pos = 'datum.start / 1000000',\n",
    "            mb_stop_pos = 'datum.stop / 1000000',\n",
    "        )\n",
    "    \n",
    "        #plot_list.append(roh_plot)\n",
    "        columns.append(roh_plot)\n",
    "    \n",
    "    return alt.hconcat(*columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_all_trio_roh(df):\n",
    "    # Find sequenced trios\n",
    "    sequenced_animals = df[\"sample\"].unique()\n",
    "    trios = colonies.filter(\n",
    "        pl.col(\"Id\").is_in(sequenced_animals) & pl.col(\"Sire\").is_in(sequenced_animals) & pl.col(\"Dam\").is_in(sequenced_animals)\n",
    "    ).select(\"Id\", \"Sire\", \"Dam\").sort(\"Id\")\n",
    "\n",
    "    # Pull out specific trio\n",
    "    rows = []\n",
    "    for trio in trios.transpose().select([f\"column_{num}\" for num in range(50,75)]):\n",
    "        trio_df = roh_df.filter(\n",
    "            pl.col(\"sample\").is_in(trio)\n",
    "        ).select(\"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Colony\")\n",
    "        \n",
    "        rows.append(plot_trio_roh(trio_df))\n",
    "    return alt.vconcat(*rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dataframe for subsequent plots\n",
    "\n",
    "chrom_df = pl.DataFrame({\n",
    "    \"chrom\": chromosomes,\n",
    "    \"chrom_len\": chromosome_lengths\n",
    "})\n",
    "genome_length = sum(chromosome_lengths)\n",
    "\n",
    "total_len_roh = roh_df.group_by(\"sample\", \"Colony\", \"Interval\").agg(pl.sum(\"length\")).sort(\"sample\").with_columns(\n",
    "    pl.col(\"length\").truediv(genome_length).alias(\"froh\")\n",
    "# ).with_columns(\n",
    "#     # Add seq info for splitting by seq\n",
    "#     seq = pl.col(\"sample\").str.slice(0, 3)\n",
    "# )\n",
    "# .with_columns(\n",
    "#     pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    ").group_by(\"sample\").agg(pl.last(\"froh\"), pl.last(\"Colony\"), pl.last(\"Interval\")).with_columns(\n",
    "    # Pull out the WGS over WES and select first\n",
    "    #pl.col(\"sample\").list.sort(descending=True).list.first(),\n",
    "    #pl.col(\"froh\").list.sort(descending=True).list.first(),\n",
    ").with_columns(\n",
    "    cohort = pl.concat_str([pl.col(\"Colony\"), pl.col(\"Interval\")], separator=\"_\")\n",
    "    #pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3),\n",
    ").filter(\n",
    "    ~pl.col(\"Interval\").is_null()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boxplot of fROH over intervals\n",
    "\n",
    "boxplot = alt.Chart(total_len_roh).mark_boxplot().encode(\n",
    "    alt.X(\"Interval:N\", title=\"Cohort\",\n",
    "          # TODO: Generalize this\n",
    "        sort=['Founders', 'pre-2003', '2003-2005', '2006-2008', '2009-2011', '2012-2014', '2015-2017', '2018-2020', 'Founders2', 'Merged']),\n",
    "    alt.Y(\"froh:Q\", title=\"fROH\"),\n",
    "    #alt.Column(\"seq\", title=\"Sequence Type\"),\n",
    "    color=alt.Color(\"Colony:N\"),\n",
    "    # tooltip=[\n",
    "    #     alt.Tooltip()\n",
    "    # ]\n",
    ").properties(\n",
    "    #width=100\n",
    "    title=\"fROH Across Cohorts\"\n",
    ")\n",
    "\n",
    "boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Violin plot (without error bars or means)\n",
    "\n",
    "violin = alt.Chart(total_len_roh).transform_density(\n",
    "    'froh',\n",
    "    as_=['froh', 'density'],\n",
    "    extent=[0, 0.05],\n",
    "    groupby=['Interval', 'Colony']\n",
    ").mark_area(orient='horizontal').encode(\n",
    "    alt.X(\"density:Q\").stack('center')\n",
    "        .impute(None)\n",
    "        .title(None)\n",
    "        .axis(labels=False, values=[0], grid=False, ticks=True),\n",
    "    alt.Y(\"froh:Q\", title=\"fROH\"),\n",
    "    alt.Column(\"Interval:N\", title=\"Cohort\",\n",
    "          # TODO: Generalize this\n",
    "        sort=['Founders', 'pre-2003', '2003-2005', '2006-2008', '2009-2011', '2012-2014', '2015-2017', '2018-2020', 'Founders2', 'Merged'],   \n",
    "        ).spacing(0).header(titleOrient='bottom', labelOrient='bottom', labelPadding=0),\n",
    "    color=alt.Color(\"Colony:N\"),\n",
    ").properties(\n",
    "    width=60,\n",
    "    title=\"fROH Across Cohorts\"\n",
    ")\n",
    "\n",
    "\n",
    "violin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Violin plot but with error bars and means\n",
    "# Has to be more complicated in order to construct a layered chart that also is faceted\n",
    "# Variations on the graph can be made my adjusting the scale on the alt.Y of the violin plot\n",
    "\n",
    "violin = alt.Chart().transform_density(\n",
    "    'froh',\n",
    "    as_=['froh', 'density'],\n",
    "    #extent=[0, 0.06],\n",
    "    groupby=['Interval', 'Colony']\n",
    ").mark_area(orient='horizontal').encode(\n",
    "    alt.X(\"density:Q\").stack('center')\n",
    "        .impute(None)\n",
    "        .title(None)\n",
    "        .axis(labels=False, values=[0], grid=False, ticks=True)\n",
    "        .scale(nice=False,zero=False),\n",
    "    #alt.X(\"Interval\"),\n",
    "    #alt.Y(\"froh:Q\", title=x_title).axis(range=[0, 0.05]),\n",
    "    alt.Y(\"froh:Q\", title='fROH'),#.scale(domain=[0, 0.05]),\n",
    "    # alt.Column(\"Interval:N\", title=\"Cohort\",\n",
    "    #       # TODO: Generalize this\n",
    "    #     sort=['Founders', 'pre-2003', '2003-2005', '2006-2008', '2009-2011', '2012-2014', '2015-2017', '2018-2020', 'Founders2', 'Merged'],   \n",
    "    #     ).spacing(0).header(titleOrient='bottom', labelOrient='bottom', labelPadding=0),\n",
    "    color=alt.Color(\"Colony:N\"),\n",
    ").properties(\n",
    "    width=60\n",
    ")\n",
    "\n",
    "error = alt.Chart().mark_errorbar(extent='ci').encode(\n",
    "    #alt.X('Interval', title=None),\n",
    "    alt.Y('froh', title='fROH')\n",
    "    )\n",
    "\n",
    "mean = alt.Chart().mark_circle(color='black').encode(\n",
    "    #alt.X('Interval', title=None),\n",
    "    alt.Y('mean(froh)', title='fROH')\n",
    "    )\n",
    "\n",
    "alt.layer(violin, error, mean, data=total_len_roh).facet(\n",
    "    #column='Interval'\n",
    "    alt.Column('Interval',\n",
    "        sort=['Founders', '2003-2005', '2006-2008', '2009-2011', '2012-2014', '2015-2017', '2018-2020', 'Founders2', 'Merged'],\n",
    "        header=alt.Header(labelOrient='bottom', titleOrient='bottom', labelPadding=0, title='Cohort')\n",
    "    )\n",
    ").resolve_scale(x=alt.ResolveMode(\"independent\")\n",
    ").configure_facet(\n",
    "    spacing=0,\n",
    ").configure_title(anchor='middle').properties(title=\"fROH Across Cohorts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframes for t-tests\n",
    "merged = total_len_roh.filter(\n",
    "    pl.col('Colony') == 'Merged'\n",
    ")['froh']\n",
    "\n",
    "colony1 = total_len_roh.filter(\n",
    "    (pl.col('Colony') == '1') & (pl.col('Interval') == '2018-2020')\n",
    ")['froh']\n",
    "\n",
    "colony2 = total_len_roh.filter(\n",
    "    (pl.col('Colony') == '2') & (pl.col('Interval') == 'Founders2')\n",
    ")['froh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# T-test to compare groups\n",
    "import scipy\n",
    "\n",
    "# Comare Colony1 to Merged\n",
    "scipy.stats.ttest_ind(colony1, merged)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.ttest_ind(colony2, merged)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
