{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import seaborn as sns\n",
    "\n",
    "# Set chromosome\n",
    "#chromosomes = [\"1\", \"3\", \"5\",\"18\"]\n",
    "chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "\n",
    "# Define path to files below\n",
    "home = '/master/abagwell'\n",
    "#samples_list = f\"{home}/variant-analysis/resources/rhesus/samples/runs.WGS.U42.list\"\n",
    "samples_list = f\"{home}/variant-analysis/resources/rhesus/samples/runs.WGS_WES.U42.list\"\n",
    "contig_length = f\"{home}/variant-analysis/results/rhesus/relatedness/roh/contig_lengths.tsv\"\n",
    "\n",
    "# Create list of sample names\n",
    "sample_names = []\n",
    "with open(samples_list, \"r\") as f:\n",
    "    for sample in f:\n",
    "        sample_names.append(\"_\".join(sample.strip().split(\"_\")[0:2]))\n",
    "\n",
    "# Create contig lists\n",
    "contigs = pd.read_table(contig_length, names=[\"chr\", \"length\"], index_col=\"chr\")\n",
    "contig_lengths = contigs.to_dict()[\"length\"]\n",
    "chrom_length = contig_lengths[\"1\"] #contigs_lengths.iloc[0]\n",
    "\n",
    "# Find total lengths of chromosomes\n",
    "total_length = 0\n",
    "for contig, length in contig_lengths.items():\n",
    "    if contig in chromosomes:\n",
    "        total_length += length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy.cluster import hierarchy\n",
    "from scipy.spatial.distance import pdist\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Must add individuals in same order as in matrix\n",
    "kinship_samples = []\n",
    "kinship_csv = \"/master/abagwell/variant-analysis/resources/rhesus/pedigree/kinship_matrix.csv\"\n",
    "kinship_matrix = np.genfromtxt(kinship_csv, delimiter=\",\")\n",
    "\n",
    "# for sample in kinship_samples:\n",
    "#     if f\"WGS{sample}\" in sample_names:\n",
    "#         sample_labels.append(f\"WGS{sample}\")\n",
    "#     elif f\"WES{sample}\" in sample_names:\n",
    "#         sample_labels.append(f\"WES{sample}\")\n",
    "#     else:\n",
    "#         print(\"ID not in samples.\")\n",
    "\n",
    "# Set identities to 1\n",
    "np.fill_diagonal(kinship_matrix, 1)\n",
    "\n",
    "# Set individuals with no kinship to a small number (since smaller kinship coefficients here mean less related)\n",
    "for series in kinship_matrix:\n",
    "    series[series == 0] = 0.0001  # Less than recorded kinship values\n",
    "\n",
    "# Make values linear with log2, make positive with absolute value, and then subtract to to set identical pairs to 0\n",
    "# Values will then be zero if identical and higher numbers are more distantly related. These grow linearly\n",
    "kinship_matrix = np.absolute(np.log2(kinship_matrix))\n",
    "print(kinship_matrix)\n",
    "\n",
    "# Plot dendrogram\n",
    "sns.set_theme()\n",
    "hierarchy.set_link_color_palette(['m', 'c', 'crimson', 'orange'])\n",
    "Z = hierarchy.linkage(pdist(kinship_matrix, \"euclidean\"), 'single')\n",
    "plt.figure(figsize=(8,15))\n",
    "dn = hierarchy.dendrogram(Z, orientation='left', labels=kinship_samples, color_threshold=20, above_threshold_color='black')\n",
    "\n",
    "# Get colors and order for ROHs\n",
    "colors = dn[\"leaves_color_list\"]\n",
    "leaves = dn[\"ivl\"]  #dn[\"leaves\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only run this block if there is no grouping by kinship\n",
    "leaves = sorted(list(set([sample[3:] for sample in sample_names])))\n",
    "leaves = [leaf.split(\"/\")[1] for leaf in leaves]  # Requried for when the samples include batch name separated by \"/\"\n",
    "leaves = [\"_\".join(leaf.split(\"_\")[0:2]) for leaf in leaves]\n",
    "sample_names = [sample.split(\"/\")[1] for sample in sample_names]\n",
    "colors = ['teal' for leaf in leaves]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Attempting with polars instead of pandas\n",
    "chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "chromosome_lengths = [223_616_942, 196_197_964, 185_288_947, 169_963_040, 187_317_192, 179_085_566, 169_868_564, 145_679_320, 134_124_166, 99_517_758, 133_066_086, 130_043_856, 108_737_130, 128_056_306, 113_283_604, 79_627_064, 95_433_459, 74_474_043, 58_315_233, 77_137_495]  # For Mmul_10\n",
    "chrom_df = pl.DataFrame({\n",
    "    \"chrom\": chromosomes,\n",
    "    \"chrom_len\": chromosome_lengths\n",
    "})\n",
    "\n",
    "froh_dfs = []\n",
    "for idx, chrom in enumerate(chromosomes):\n",
    "    froh_file = f\"{home}/variant-analysis/results/rhesus/relatedness/roh/merged/U42_WGS_WES.SNP.chr{chrom}.froh_poisson.pickle\"\n",
    "    froh_df = pl.from_pandas(pd.read_pickle(froh_file))\n",
    "    froh_dfs.append(froh_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find fROH\n",
    "froh_df = pl.concat(froh_dfs).join(chrom_df, on=\"chrom\").with_columns(\n",
    "    # Find total length of ROHs (per chromosome)\n",
    "    pl.col(\"chrom_len\").mul(pl.col(\"froh\")).alias(\"roh_len\")\n",
    ").group_by(\"sample\"\n",
    "# Sum lengths from chromosomes\n",
    ").agg(pl.sum(\"roh_len\")\n",
    ").with_columns(\n",
    "    # Find fraction of genomes in ROH\n",
    "    pl.col(\"roh_len\").truediv(pl.lit(sum(chromosome_lengths))).alias(\"froh\")\n",
    ")\n",
    "froh_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colony assignments\n",
    "colonies_file = \"/master/abagwell/variant-analysis/resources/rhesus/pop/colonies.tsv\"\n",
    "colonies = pl.read_csv(colonies_file, separator=\"\\t\", infer_schema_length=None)\n",
    "\n",
    "froh_df_by_colony = froh_df.with_columns(\n",
    "    pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"id\"),\n",
    ").join(colonies, how=\"left\", left_on=\"id\", right_on=\"Id\").drop_nulls()\n",
    "#froh_df_by_colony\n",
    "\n",
    "# merged = colony_demographics.join(Q, how=\"left\", left_on=\"Id\", right_on=\"#sample\"\n",
    "# ).group_by(\"Year\", \"Colony\").agg(pl.count(\"Id\").alias(\"Count\"), pl.mean(\"Indian\"), pl.mean(\"Chinese\")).drop_nulls()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot Altair boxplot\n",
    "import altair as alt\n",
    "\n",
    "alt.Chart(froh_df_by_colony).mark_boxplot().encode(\n",
    "    alt.X(\"Colony:N\"),\n",
    "    alt.Y(\"froh:Q\", title=\"fROH\"),\n",
    "    color=alt.Color(\"Colony:N\"),\n",
    ").properties(\n",
    "    #width=100\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Seaborn boxplot to show percentage of genome in ROH\n",
    "\n",
    "g = sns.catplot(data=froh_df_by_colony, y=\"froh\", x=\"Colony\", hue=\"Colony\", kind=\"box\", aspect=0.6).set(ylabel=\"Fraction of Genome\", title=\"Extent of ROHs in\\nRhesus Macaque Populations\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "froh_pandas = froh_df_by_colony.to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "froh_pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "froh_pandas[froh_pandas[\"id\"] == \"30009\"].values[0][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate figure with dendrogram and fROH\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize=(10, 30)) #height_ratios=[height, 2])\n",
    "#fig.suptitle(f\"FROH\")\n",
    "fig.tight_layout()\n",
    "plt.subplots_adjust(hspace=0.1)\n",
    "plt.subplots_adjust(wspace=0.27)\n",
    "\n",
    "dnn = hierarchy.dendrogram(Z, ax=ax[0], orientation='left', labels=kinship_samples, no_labels=True, color_threshold=20, above_threshold_color='black')\n",
    "\n",
    "#percent_froh = combined_froh[combined_froh[\"index\"] == sample].values[0][1] / total_length\n",
    "\n",
    "for indiv, color in zip(leaves, colors):\n",
    "    ax[1].barh(indiv, froh_pandas[froh_pandas[\"id\"] == indiv].values[0][2] / total_length, color=color)\n",
    "ax[1].set_xlabel(\"fROH\")\n",
    "plt.margins(y=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect invidual ids to submit in TAC to obtain kinship matrix\n",
    "for sample in sorted(list(set([sample.split(\"/\")[-1].split(\"_\")[0][3:] for sample in sample_names]))):\n",
    "    print(sample, end=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# regions_file = \"/master/abagwell/variant-analysis/results/rhesus/coverage/common_WES_0.5_loci.bed\"\n",
    "# pl.read_csv(regions_file, separator=\"\\t\", has_header=False, new_columns=[\"chrom\", \"start\", \"stop\"], dtypes=[pl.String, pl.Int32, pl.Int32])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Altair. Using output from `bcftools roh`\n",
    "import altair as alt\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "#chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "#chromosome_lengths = [223_616_942, 196_197_964, 185_288_947, 169_963_040, 187_317_192, 179_085_566, 169_868_564, 145_679_320, 134_124_166, 99_517_758, 133_066_086, 130_043_856, 108_737_130, 128_056_306, 113_283_604, 79_627_064, 95_433_459, 74_474_043, 58_315_233, 77_137_495]  # For Mmul_10\n",
    "\n",
    "chromosomes = [\"1\", \"2\"]\n",
    "chromosome_lengths = [223_616_942, 196_197_964]  # For Mmul_10\n",
    "\n",
    "home = '/master/abagwell'\n",
    "samples_list = f\"{home}/variant-analysis/resources/rhesus/samples/runs.WGS.U42.list\"\n",
    "contig_length = f\"{home}/variant-analysis/results/rhesus/relatedness/roh/contig_lengths.tsv\"\n",
    "\n",
    "\n",
    "# Read colony info\n",
    "colonies_file = \"/master/abagwell/variant-analysis/resources/rhesus/pop/colonies.tsv\"\n",
    "colonies = pl.read_csv(colonies_file, separator=\"\\t\", infer_schema_length=None)\n",
    "\n",
    "# Create list of sample names\n",
    "sample_names = []\n",
    "with open(samples_list, \"r\") as f:\n",
    "    for sample in f:\n",
    "        sample_names.append(\"_\".join(sample.strip().split(\"/\")[1].split(\"_\")[0:2]))\n",
    "\n",
    "\n",
    "plot_list = []\n",
    "columns = []\n",
    "for idx, chrom in enumerate(chromosomes):\n",
    "    column_plots = []\n",
    "    # Only keep y-axis labels on leftmost plot\n",
    "    if idx == 0:\n",
    "        axis = alt.Axis(tickSize=0)\n",
    "    else:\n",
    "        axis = None\n",
    "    # Output file from `bcftools roh`\n",
    "    rg_file = f\"/master/abagwell/variant-analysis/results/rhesus/roh/SHAPEIT5_WGS/U42_WGS_WES.SNP.chr{chrom}.RG.roh\"\n",
    "    rg_df = pl.read_csv(rg_file, comment_prefix=\"#\", has_header=False, separator=\"\\t\",\n",
    "        new_columns=[\"tag\", \"sample\", \"chrom\", \"start\", \"end\", \"length\", \"num_of_markers\", \"quality\"],\n",
    "        dtypes=[pl.String, pl.String, pl.String, pl.Int32, pl.Int32, pl.Int32, pl.Int32, pl.Float64]\n",
    "    ).filter(\n",
    "        pl.col(\"length\") > 5_000_000\n",
    "    ).with_columns(\n",
    "        pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    "    ).join(colonies, how=\"left\", left_on=\"animal\", right_on=\"Id\")\n",
    "\n",
    "    # Add dummy line for each sample to guarantee that it will show up even if it has no ROH\n",
    "    # dummies = []\n",
    "    # for sample in sample_names:\n",
    "    #     rg_df_with_dummy = pl.concat([rg_df, rg_df.group_by(\"tag\", \"sample\", \"chrom\").agg().with_columns(\n",
    "    #         # Add back columns with constant values\n",
    "    #         pl.lit(621085).cast(pl.Int32).alias(\"start\"),\n",
    "    #         pl.lit(2599489).cast(pl.Int32).alias(\"end\"),\n",
    "    #         pl.lit(0).cast(pl.Int32).alias(\"length\"),\n",
    "    #         pl.lit(0).cast(pl.Int32).alias(\"num_of_markers\"),\n",
    "    #         pl.lit(0).cast(pl.Float64).alias(\"quality\"),\n",
    "    #     )])\n",
    "\n",
    "    rg_pandas = rg_df.to_pandas()\n",
    "    for sample in sample_names:\n",
    "       rg_pandas.loc[len(rg_pandas)] = [\"RG\", sample, chrom, 0, 0, 0, 0, 0, \"\", 1]\n",
    "    rg_df_with_dummy = pl.from_pandas(rg_pandas)\n",
    "\n",
    "\n",
    "    roh_plot = alt.Chart(rg_df_with_dummy).mark_bar().encode(\n",
    "        alt.X(\"mb_start_pos:Q\").title([\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),\n",
    "        alt.X2(\"mb_stop_pos:Q\"), #.title(\"Stop\"),\n",
    "        alt.Y(\"sample\", axis=axis).title(\"Sample\"),\n",
    "        color = alt.Color(\"Colony:N\"),  # .scale(range=['#1f77b4', 'orange', '#9F2B68'])\n",
    "        #color = \"orange\",\n",
    "        tooltip=[\n",
    "            alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "            alt.Tooltip(\"start\", title=\"Start (bp)\"),\n",
    "            alt.Tooltip(\"end\", title=\"End (bp)\"),\n",
    "            alt.Tooltip(\"length\", title=\"Length\"),\n",
    "        ]\n",
    "    ).properties(\n",
    "        width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "        height=alt.Step(10),\n",
    "        #width=300, # For when only displaying one chromosome\n",
    "        title=[\"RoH\", f\"chr{chrom}\"]\n",
    "    ).transform_calculate(\n",
    "        mb_start_pos = 'datum.start / 1000000',\n",
    "        mb_stop_pos = 'datum.end / 1000000',\n",
    "    )\n",
    "    plot_list.append(roh_plot)\n",
    "    column_plots.append(roh_plot)\n",
    "\n",
    "    # Create df for bar plot\n",
    "    bar_df = rg_df.with_columns(\n",
    "        pl.int_ranges(pl.col(\"start\"), pl.col(\"end\"), 1_000).alias(\"position\")\n",
    "        ).explode(\"position\").group_by(\"chrom\", \"position\", \"Colony\").agg(pl.len())\n",
    "\n",
    "    # Create bar plot\n",
    "    if idx == 0:\n",
    "        axis = alt.Axis(labels=False, title=None)\n",
    "    else:\n",
    "        axis = alt.Axis(labels=False, ticks=False, title=\"\")\n",
    "    bar_plot = alt.Chart(bar_df).mark_bar(size=1).encode(\n",
    "        alt.X(\"mb_pos:Q\").title([\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),  ## Modified here last\n",
    "        alt.Y(\"len\").title(\"Count\").scale(domainMin=0, domainMax=bar_df[\"len\"].max()), #axis=axis\n",
    "        alt.Row(\"Colony:N\", title=None),\n",
    "        color=alt.Color(\"Colony:N\"),\n",
    "        ).properties(\n",
    "            width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "            height=70,\n",
    "            #width=300, # For when only displaying one chromosome\n",
    "            title=[\"RoH\", f\"chr{chrom}\"]\n",
    "        ).transform_calculate(\n",
    "            mb_pos = 'datum.position / 1000000',\n",
    "        )\n",
    "\n",
    "    column_plots.append(bar_plot)\n",
    "    columns.append(alt.vconcat(*column_plots))\n",
    "\n",
    "\n",
    "\n",
    "# (plot_list[0] | plot_list[1] | plot_list[2] | plot_list[16] | plot_list[19]).properties(\n",
    "#     title=\"Runs of Homozygosity\",\n",
    "# )\n",
    "alt.hconcat(*columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_len_roh = rg_df_with_dummy.group_by(\"sample\").agg(pl.sum(\"length\")).sort(\"sample\")\n",
    "\n",
    "total_roh = alt.Chart(total_len_roh).mark_bar().encode(\n",
    "    alt.X(\"length\"),\n",
    "    alt.Y(\"sample\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bar_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graph",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "de4a875ba73fa07c7c0f7a1dd2e6901433a766d080dc9d908c0725e19addd1ed"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
