{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge ROH files and filter\n",
    "# Altair. Using output from `scikit-allel`\n",
    "import altair as alt\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "from datetime import date\n",
    "\n",
    "# pedigree_file = \"/master/abagwell/variant-analysis/resources/rhesus/pedigree/Demographics_2024-04-17_10-28-20.tsv\"\n",
    "# date_of_pedigree = date(2024, 4, 17)\n",
    "# dates = pl.read_csv(pedigree_file, separator=\"\\t\", infer_schema_length=None, columns=[\"Id\", \"Most Recent Acq\"]).with_columns(\n",
    "#     #pl.col(\"Account Description\").str.contains(\"eserved for breeding\").alias(\"is_reserved_for_breeding\"),\n",
    "#     #pl.col(\"Account Description\").str.contains(\"reeder\").alias(\"is_breeder\"),\n",
    "# ).with_columns( # Change str to date type\n",
    "#     date = pl.col(\"Most Recent Acq\").str.to_datetime(\"%m-%d-%Y %H:%M\"),\n",
    "# ).drop(\"Most Recent Acq\").with_columns(  # Set death date of living animals to date of pedigree\n",
    "#     #pl.col(\"Date of Death\").fill_null(date_of_pedigree),\n",
    "# )\n",
    "\n",
    "\n",
    "\n",
    "# Read colony info\n",
    "colonies_file = \"/master/abagwell/variant-analysis/resources/rhesus/pop/colonies.tsv\"\n",
    "colonies = pl.read_csv(colonies_file, separator=\"\\t\", infer_schema_length=None)#.join(dates, on=\"Id\", how=\"left\")\n",
    "\n",
    "chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "chromosome_lengths = [223_616_942, 196_197_964, 185_288_947, 169_963_040, 187_317_192, 179_085_566, 169_868_564, 145_679_320, 134_124_166, 99_517_758, 133_066_086, 130_043_856, 108_737_130, 128_056_306, 113_283_604, 79_627_064, 95_433_459, 74_474_043, 58_315_233, 77_137_495]  # For Mmul_10\n",
    "#chromosomes = [\"1\", \"2\"]\n",
    "#chromosome_lengths = [223_616_942, 196_197_964]  # For Mmul_10\n",
    "\n",
    "\n",
    "roh_dfs = []\n",
    "for chrom in chromosomes:\n",
    "    roh_file = f\"/master/abagwell/variant-analysis/results/rhesus/roh/scikit-allel/with_is_accessible/U42_WGS_WES.SNP.chr{chrom}.roh_poisson.tsv\"\n",
    "    roh_df = pl.read_csv(roh_file, comment_prefix=\"#\", has_header=True, separator=\"\\t\",\n",
    "        dtypes=[pl.String, pl.String, pl.Int32, pl.Int32, pl.Int32, pl.Boolean]\n",
    "    ).filter(\n",
    "        # Keep only large ROHs\n",
    "        pl.col(\"length\") > 1_000_000\n",
    "    ).with_columns(\n",
    "        pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    "    ).join(colonies, how=\"left\", left_on=\"animal\", right_on=\"Id\").drop(\"tag\")\n",
    "    \n",
    "    roh_dfs.append(roh_df)\n",
    "\n",
    "roh_df = pl.concat(roh_dfs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read sample names from fROH file\n",
    "froh_file = f\"/master/abagwell/variant-analysis/results/rhesus/roh/scikit-allel/with_is_accessible/U42_WGS_WES.SNP.chr{chrom}.froh_poisson.tsv\"\n",
    "froh_df = pl.read_csv(froh_file, comment_prefix=\"#\", has_header=True, separator=\"\\t\",\n",
    "    new_columns=[\"sample\", \"chrom\", \"froh\"],\n",
    "    dtypes=[pl.String, pl.String, pl.Float64]\n",
    ").with_columns(\n",
    "    pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "sample_names = list(froh_df.unique(\"sample\").sort(\"sample\")[\"sample\"])\n",
    "\n",
    "# Create dummy lines\n",
    "#dummies = roh_df.group_by(\"tag\", \"sample\", \"chrom\").agg(pl.first(\"*\"))\n",
    "\n",
    "dummies = pl.DataFrame([\n",
    "    pl.Series(\"sample\", sample_names, dtype=pl.String),\n",
    "]).join(pl.DataFrame([pl.Series(\"chrom\", chromosomes, dtype=pl.String),]), how=\"cross\").with_columns(\n",
    "    pl.lit(0).alias(\"start\"),\n",
    "    pl.lit(0).alias(\"stop\"),\n",
    "    pl.lit(0).alias(\"length\"),\n",
    "    pl.lit(False).alias(\"is_marginal\"),\n",
    ").with_columns(\n",
    "    pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    ").join(colonies, how=\"left\", left_on=\"animal\", right_on=\"Id\")\n",
    "\n",
    "roh_df_with_dummy = pl.concat([roh_df, dummies])\n",
    "\n",
    "# Keep only 1 instance of each animal\n",
    "no_duplicates = roh_df_with_dummy.group_by(\"animal\").agg(\"sample\").with_columns(\n",
    "    # Pull out the WGS over WES and select first\n",
    "    pl.col(\"sample\").list.sort(descending=True).list.first(),\n",
    ").with_columns(\n",
    "    #pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3),\n",
    ")\n",
    "roh_df_no_duplicates = roh_df_with_dummy.join(no_duplicates, on=\"sample\", how=\"semi\"\n",
    ").with_columns(\n",
    "    # Keep only animal name\n",
    "    sample = pl.col(\"animal\")\n",
    ")\n",
    "\n",
    "# Alternatively, keep only duplicates\n",
    "only_duplicates = roh_df_with_dummy.group_by(\"animal\").agg(\"sample\").with_columns(\n",
    "    # Pull out the WGS over WES and select first\n",
    "    pl.col(\"sample\").list.unique()#list.sort(descending=True).list.first(),\n",
    ").filter(\n",
    "    pl.col(\"sample\").list.len() > 1\n",
    ").explode(\"sample\")\n",
    "roh_df_only_duplicates = roh_df_with_dummy.join(only_duplicates, on=\"sample\", how=\"semi\"\n",
    ").with_columns(\n",
    "    # Keep only animal name\n",
    "    #sample = pl.col(\"animal\")\n",
    ")\n",
    "\n",
    "# Set which df to use: roh_df_no_duplicates or roh_df_only_duplicates\n",
    "#roh_df = roh_df_only_duplicates\n",
    "roh_df = roh_df_no_duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Altair. Using output from `bcftools roh`\n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "def roh_plot_colony(roh_df, colony, color):\n",
    "\n",
    "    #chromosomes = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\"]\n",
    "    #chromosome_lengths = [223_616_942, 196_197_964, 185_288_947, 169_963_040, 187_317_192, 179_085_566, 169_868_564, 145_679_320, 134_124_166, 99_517_758, 133_066_086, 130_043_856, 108_737_130, 128_056_306, 113_283_604, 79_627_064, 95_433_459, 74_474_043, 58_315_233, 77_137_495]  # For Mmul_10\n",
    "    # chromosomes = [\"1\", \"2\"]\n",
    "    # chromosome_lengths = [223_616_942, 196_197_964]  # For Mmul_10\n",
    "\n",
    "\n",
    "    home = '/master/abagwell'\n",
    "    contig_length = f\"{home}/variant-analysis/results/rhesus/relatedness/roh/contig_lengths.tsv\"\n",
    "\n",
    "    # Find max fROH in Mb across all colony groups so that each plot will have the same axis\n",
    "    max_froh = roh_df.group_by(\"sample\").agg(pl.sum(\"length\"))[\"length\"].max() / 1_000_000\n",
    "\n",
    "\n",
    "    # Subset df\n",
    "    roh_df = roh_df.filter(\n",
    "        # Keep only members of specified colony\n",
    "        pl.col(\"Colony\") == colony\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "    bar_df = roh_df.filter(\n",
    "        # Remove dummy variables\n",
    "        pl.col(\"length\") > 0\n",
    "    ).with_columns(\n",
    "        pl.int_ranges(pl.col(\"start\"), pl.col(\"stop\"), 1_000_000).alias(\"position\")\n",
    "        ).explode(\"position\").group_by(\"chrom\", \"position\", \"Colony\").agg(pl.len())\n",
    "\n",
    "    max_bar_height = bar_df[\"len\"].max()\n",
    "\n",
    "\n",
    "    plot_list = []\n",
    "    columns = []\n",
    "    for idx, chrom in enumerate(chromosomes):\n",
    "        column_plots = []\n",
    "        # Only keep y-axis labels on leftmost plot\n",
    "        if idx == 0:\n",
    "            axis = alt.Axis()\n",
    "        else:\n",
    "            axis = None\n",
    "        # Info from `bcftools roh`\n",
    "        chrom_roh_df = roh_df.filter(\n",
    "            # Keep only large ROHs\n",
    "            #pl.col(\"length\") > 2_000_000\n",
    "            pl.col(\"chrom\") == chrom\n",
    "        )\n",
    "\n",
    "        roh_plot = alt.Chart(chrom_roh_df).mark_bar().encode(\n",
    "            alt.X(\"mb_start_pos:Q\").title([\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),\n",
    "            alt.X2(\"mb_stop_pos:Q\"), #.title(\"Stop\"),\n",
    "            alt.Y(\"sample:O\", axis=axis, title=\"Sample\"), # sort=alt.EncodingSortField(field='date', order='ascending')\n",
    "            #color = alt.Color(\"Colony:N\"),  # .scale(range=['#1f77b4', 'orange', '#9F2B68'])\n",
    "            color = alt.value(color),\n",
    "            #order=alt.Order('animal:O', sort='ascending'),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "                alt.Tooltip(\"start\", title=\"Start (bp)\"),\n",
    "                alt.Tooltip(\"stop\", title=\"End (bp)\"),\n",
    "                alt.Tooltip(\"length\", title=\"Length\"),\n",
    "            ]\n",
    "        ).properties(\n",
    "            width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "            height=alt.Step(10),\n",
    "            #width=300, # For when only displaying one chromosome\n",
    "            title=[\"RoH\", f\"chr{chrom}\"]\n",
    "        ).transform_calculate(\n",
    "            mb_start_pos = 'datum.start / 1000000',\n",
    "            mb_stop_pos = 'datum.stop / 1000000',\n",
    "        )\n",
    "        plot_list.append(roh_plot)\n",
    "        column_plots.append(roh_plot)\n",
    "\n",
    "        chrom_bar_df = bar_df.filter(\n",
    "            pl.col(\"chrom\") == chrom\n",
    "        )\n",
    "\n",
    "\n",
    "        # Create bar plot\n",
    "        if idx == 0:\n",
    "            axis = alt.Axis()\n",
    "        else:\n",
    "            axis = alt.Axis(labels=False, title=None)  #ticks=False,\n",
    "        bar_plot = alt.Chart(chrom_bar_df).mark_bar(size=1).encode(\n",
    "            alt.X(\"mb_pos:Q\").title([\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),  ## Modified here last\n",
    "            alt.Y(\"len\", axis=axis).title(\"Count\").scale(domainMin=0, domainMax=max_bar_height), #axis=axis  #, domainMax=bar_df[\"len\"].max()\n",
    "            #alt.Row(\"Colony:N\", title=None), \n",
    "            #color=alt.Color(\"Colony:N\"),\n",
    "            color=alt.value(color)\n",
    "            ).properties(\n",
    "                width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "                height=70,\n",
    "                #width=300, # For when only displaying one chromosome\n",
    "                title=[\"RoH\", f\"chr{chrom}\"]\n",
    "            ).transform_calculate(\n",
    "                mb_pos = 'datum.position / 1000000',\n",
    "            )\n",
    "\n",
    "        column_plots.append(bar_plot)\n",
    "        columns.append(alt.vconcat(*column_plots))\n",
    "\n",
    "    total_len_roh = roh_df.group_by(\"sample\").agg(pl.sum(\"length\"))#.sort(\"sample\")\n",
    "\n",
    "    total_roh = alt.Chart(total_len_roh).mark_bar().encode(\n",
    "        alt.X(\"mb_length:Q\", title=\"Total length (Mb)\").scale(domainMin=0, domainMax=max_froh, clamp=True),\n",
    "        #alt.X2(),\n",
    "        alt.Y(\"sample:O\", title=\"Sample\"), #sort=alt.EncodingSortField(field='date', order='ascending')\n",
    "        color=alt.value(color),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(\"length\", title=\"Length\"),\n",
    "        ]\n",
    "    ).properties(\n",
    "        height=alt.Step(10),\n",
    "        width=100,\n",
    "        title=\"Total ROH Length\",\n",
    "    ).transform_calculate(\n",
    "        mb_length = 'datum.length / 1000000',\n",
    "    )\n",
    "    columns.append(total_roh)\n",
    "\n",
    "    return alt.hconcat(*columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "roh_plot_colony(roh_df, 1, \"teal\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "roh_plot_colony(roh_df, 2, \"orange\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graph",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "de4a875ba73fa07c7c0f7a1dd2e6901433a766d080dc9d908c0725e19addd1ed"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
