{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To replace above block\n",
    "\n",
    "import allel\n",
    "import altair as alt\n",
    "from altair import datum\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "\n",
    "# Query VCF\n",
    "vcf = \"/master/abagwell/variant-analysis/results/rhesus/structural_variants/annotated/U42_WGS_WES.genotyped.pass.split_BNDs.vcf.gz\"\n",
    "variants = pl.from_pandas(\n",
    "    pd.DataFrame(\n",
    "        allel.read_vcf(vcf, ['variants/CHROM', 'variants/POS', 'variants/SVTYPE', 'variants/ID', 'variants/CHR2', 'variants/POS2', 'variants/CSQ'])\n",
    "    )\n",
    ")\n",
    "\n",
    "pops = pl.read_csv(\n",
    "    #\"/master/abagwell/variant-analysis/resources/rhesus/samples/pops/RPL.tsv\",\n",
    "    \"/master/abagwell/variant-analysis/resources/rhesus/samples/pops/RPL_with_males.tsv\",\n",
    "    separator=\"\\t\"\n",
    ").group_by(\"RPL_status\"\n",
    ").agg(\"sample\")\n",
    "\n",
    "\n",
    "groups = []\n",
    "for pop_id, pop_members in pops.rows():\n",
    "    callset = allel.read_vcf(vcf, ['calldata/GT'], samples=pop_members)\n",
    "    genotypes = allel.GenotypeArray(callset['calldata/GT'])\n",
    "\n",
    "    # Allele counts\n",
    "    #ac_tmp = genotypes.count_alleles()\n",
    "    ac = pl.from_pandas(pd.DataFrame(genotypes.count_alleles())).with_columns(\n",
    "        pl.lit(pop_id).alias(\"group\"),\n",
    "        pl.col(\"1\").truediv(pl.col(\"0\").add(pl.col(\"1\"))).alias(\"fraction\" + str(pop_id)),\n",
    "        pl.col(\"0\").add(pl.col(\"1\")).alias(\"AN\" + str(pop_id)),\n",
    "    ).drop(\n",
    "        [\"group\", \"0\", \"1\"]\n",
    "    )\n",
    "\n",
    "    # groups.append(\n",
    "    #     pl.concat([variants, ac], how=\"horizontal\")\n",
    "    # )\n",
    "\n",
    "    groups.append(ac)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_acs = pl.concat(groups, how=\"horizontal\")\n",
    "\n",
    "all = pl.concat([variants, all_acs], how=\"horizontal\").with_columns(\n",
    "    pl.col(\"fraction0\").sub(pl.col(\"fraction1\")).alias(\"difference\"),\n",
    "    pl.concat_str([\n",
    "        pl.col(\"variants/CHROM\"),\n",
    "        pl.lit(\":\"),\n",
    "        pl.col(\"variants/POS\"),\n",
    "    ]).alias(\"locus\"),\n",
    "    pl.col(\"variants/CSQ\").str.split(\"|\").list.get(1).alias(\"consequence\"),\n",
    "    pl.col(\"variants/CSQ\").str.split(\"|\").list.get(2).alias(\"impact\"),\n",
    "    pl.col(\"variants/CSQ\").str.split(\"|\").list.get(3).alias(\"symbol\"),\n",
    "    pl.col(\"variants/CSQ\").str.split(\"|\").list.get(4).alias(\"gene\"),\n",
    "# ).filter(\n",
    "#     (pl.col(\"difference\") > 0.28) | (pl.col(\"difference\") < -0.28)\n",
    "#).filter(\n",
    "    #(pl.col(\"variants/SVTYPE\") == \"BND\")\n",
    ").drop(\n",
    "    \"variants/CSQ\"\n",
    ").group_by(\n",
    "    pl.col(\"locus\")\n",
    ").first( # Removes rows with duplicate position, keeping only one's info\n",
    ").sort(\n",
    "    pl.col(\"difference\"), descending=True\n",
    ").with_row_count()\n",
    "\n",
    "\n",
    "# all = pl.concat(groups, how=\"vertical\").filter(\n",
    "#     pl.col(\"variants/CHROM\") == \"10\",\n",
    "# )\n",
    "\n",
    "#all.group_by(pl.col(\"impact\")).count()\n",
    "all\n",
    "#all.write_csv(\"/master/abagwell/figures/SVs_biallelic.split_breakpoints.VEP.tsv\", separator=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Examples of further filtering\n",
    "# all = all.filter(\n",
    "#     #pl.col(\"variants/CHROM\") == \"15\",\n",
    "#     (pl.col(\"difference\") > 0.2) | (pl.col(\"difference\") < -0.2)\n",
    "# ).filter(\n",
    "#     pl.col(\"variants/SVTYPE\") == \"BND\"\n",
    "# )\n",
    "\n",
    "# all = all.filter(\n",
    "#     pl.col(\"fraction0\") == 0\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all = all.sort(\"fraction1\", descending = True)\n",
    "\n",
    "#all.write_csv(\"/master/abagwell/figures/SVs_including_males_RPL_only.tsv\", separator=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatterplot of allele frequencies for RPL and non-RPL\n",
    "\n",
    "slider_0 = alt.binding_range(min=0, max=1, step=0.01, name='fraction0 ')\n",
    "selector_0 = alt.selection_point(\n",
    "    name=\"selector0\",\n",
    "    fields=['fraction0'],\n",
    "    bind=slider_0,\n",
    "    value=[{'fraction0': 1}]\n",
    ")\n",
    "slider_1 = alt.binding_range(min=0, max=1, step=0.01, name='fraction1 ')\n",
    "selector_1 = alt.selection_point(\n",
    "    name=\"selector1\",\n",
    "    fields=['fraction1'],\n",
    "    bind=slider_1,\n",
    "    value=[{'fraction1': 1}]\n",
    ")\n",
    "\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "chart = alt.Chart(all.to_arrow().to_pandas()).mark_circle().encode(\n",
    "    #alt.X(\"variants/POS:N\"),\n",
    "    #alt.X(\"row_nr:N\").title(\"Structural Variants\"),\n",
    "    #alt.X(\"fraction1\").title(\"Allele Frequency in RPL\"),\n",
    "    alt.X(\"fraction1\").title(\"Allele Frequency in RPL\"),\n",
    "    alt.Y(\"fraction0\").title(\"Allele Frequency in non-RPL\"),\n",
    "    #color = alt.Color(\"group:N\"),\n",
    "    # color=alt.condition(\n",
    "    #     alt.datum.difference > 0,\n",
    "    #     alt.value(\"steelblue\"),  # The positive color\n",
    "    #     alt.value(\"orange\")  # The negative color\n",
    "    # ),\n",
    "    color = alt.Color(\"variants/SVTYPE\", scale=alt.Scale(domain=['BND', 'DUP', 'DEL', 'INV', 'INS'])).title(\"SV Type\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"symbol\", title=\"Symbol\"),\n",
    "        alt.Tooltip(\"gene\", title=\"Gene\"),\n",
    "        alt.Tooltip(\"consequence\", title=\"Consequence\"),\n",
    "        alt.Tooltip(\"impact\", title=\"Impact\"),\n",
    "        alt.Tooltip(\"variants/SVTYPE\", title=\"SVTYPE\"),\n",
    "        alt.Tooltip(\"locus\", title=\"Locus\"),\n",
    "        alt.Tooltip(\"fraction0:O\", title=\"frequency0\"),\n",
    "        alt.Tooltip(\"fraction1:O\", title=\"frequency1\"),\n",
    "        alt.Tooltip(\"difference:Q\", title=\"Difference\"),\n",
    "    ]\n",
    ").properties(\n",
    "    title = \"Correlation of SVs to RPL vs non-RPL\",\n",
    "    # height = 600,\n",
    "    # width = 600,\n",
    ").add_params(\n",
    "    selector_0, selector_1\n",
    ").transform_filter(\n",
    "    (datum.fraction0 <= selector_0.fraction0) & (datum.fraction1 <= selector_1.fraction1) \n",
    ")\n",
    "\n",
    "chart\n",
    "#chart.save(\"/master/abagwell/figures/SV_allele_frequency.WGS_including_males_RPL_only.selection2.html\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "chart = alt.Chart(all.to_arrow().to_pandas()).mark_bar().encode(\n",
    "    #alt.X(\"variants/POS:N\"),\n",
    "    #alt.X(\"row_nr:N\").title(\"Structural Variants\"),\n",
    "    alt.X(\"locus\", sort=alt.EncodingSortField(field=\"row_nr:N\", op=\"min\", order='descending')).title(\"Locus\"),\n",
    "    alt.Y(\"difference:Q\").title(\"non-RPL freq. - RPL freq.\"),\n",
    "    #color = alt.Color(\"group:N\"),\n",
    "    # color=alt.condition(\n",
    "    #     alt.datum.difference > 0,\n",
    "    #     alt.value(\"steelblue\"),  # The positive color\n",
    "    #     alt.value(\"orange\")  # The negative color\n",
    "    # ),\n",
    "    color = alt.Color(\"variants/SVTYPE\", scale=alt.Scale(domain=['BND', 'DUP', 'DEL', 'INV', 'INS'])).title(\"SV Type\"),\n",
    "    tooltip=[\n",
    "        #alt.Tooltip(\"0:O\", title=\"Ref AC\"),\n",
    "        #alt.Tooltip(\"1:O\", title=\"Alt AC\"),\n",
    "        alt.Tooltip(\"non_AN:O\", title=\"non_AN\"),\n",
    "        alt.Tooltip(\"RPL_AN:O\", title=\"RPL_AN\"),\n",
    "        alt.Tooltip(\"variants/SVTYPE\", title=\"SVTYPE\"),\n",
    "        alt.Tooltip(\"locus\", title=\"Locus\"),\n",
    "        alt.Tooltip(\"difference:Q\", title=\"Difference\"),\n",
    "    ]\n",
    ").properties(\n",
    "    title = \"SVs Most Correlated to RPL or non-RPL\",\n",
    "    width = 600,\n",
    ")\n",
    "\n",
    "chart #.save(\"/master/abagwell/figures/SV4.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "without_QNVO = all.filter(\n",
    "    ~pl.col(\"variants/CHROM\").str.contains(\"QNVO\")\n",
    ")\n",
    "\n",
    "chart = alt.Chart(without_QNVO.to_arrow().to_pandas()).mark_bar().encode(\n",
    "    #alt.X(\"variants/POS:N\"),\n",
    "    #alt.X(\"row_nr:N\").title(\"Structural Variants\"),\n",
    "    alt.X(\"locus\", sort=alt.EncodingSortField(field=\"row_nr:N\", op=\"count\", order='descending')).title(\"Locus\"),\n",
    "    alt.Y(\"difference:Q\").title(\"non-RPL freq. - RPL freq.\"),\n",
    "    #color = alt.Color(\"group:N\"),\n",
    "    # color=alt.condition(\n",
    "    #     alt.datum.difference > 0,\n",
    "    #     alt.value(\"steelblue\"),  # The positive color\n",
    "    #     alt.value(\"orange\")  # The negative color\n",
    "    # ),\n",
    "    color = alt.Color(\"variants/SVTYPE\", scale=alt.Scale(domain=['BND', 'DUP', 'DEL', 'INS'])).title(\"SV Type\"),\n",
    "    tooltip=[\n",
    "        #alt.Tooltip(\"0:O\", title=\"Ref AC\"),\n",
    "        #alt.Tooltip(\"1:O\", title=\"Alt AC\"),\n",
    "        #alt.Tooltip(\"AC:O\", title=\"Total AC\"),\n",
    "        alt.Tooltip(\"variants/SVTYPE\", title=\"SVTYPE\"),\n",
    "        alt.Tooltip(\"locus\", title=\"Locus\"),\n",
    "        alt.Tooltip(\"difference:Q\", title=\"Difference\"),\n",
    "    ]\n",
    ").properties(\n",
    "    title = \"SVs Most Correlated to RPL or non-RPL\",\n",
    "    width = 500,\n",
    ")\n",
    "\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "chart = alt.Chart(all.to_arrow().to_pandas()).mark_circle().encode(\n",
    "    alt.X(\"variants/POS\"),\n",
    "    alt.Y(\"fraction:Q\").title(\"Fraction\"),\n",
    "    color = alt.Color(\"group:N\"),\n",
    "    tooltip=[\n",
    "        alt.Tooltip(\"0:O\", title=\"Ref AC\"),\n",
    "        alt.Tooltip(\"1:O\", title=\"Alt AC\"),\n",
    "        alt.Tooltip(\"AC:O\", title=\"Total AC\"),\n",
    "        alt.Tooltip(\"variants/SVTYPE\", title=\"SVTYPE\"),\n",
    "        alt.Tooltip(\"variants/POS\", title=\"Position\"),\n",
    "        alt.Tooltip(\"fraction:Q\", title=\"Fraction Alt\"),\n",
    "    ]\n",
    ").properties(\n",
    "    title = \"SVs across Rhesus\",\n",
    "    width = 10000,\n",
    ")\n",
    "\n",
    "chart"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
