{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "import polars as pl\n",
    "import yaml\n",
    "\n",
    "# Read config file. User should update this file path as needed\n",
    "configfile = \"/master/abagwell/workspace/github_project/variant-analysis/config/marmoset.yaml\"\n",
    "with open(configfile, 'r') as file:\n",
    "    config = yaml.safe_load(file)\n",
    "\n",
    "# Load colors\n",
    "colors = pl.read_csv(config[\"colors\"], separator=\"\\t\", comment_prefix='#', schema_overrides={\"Cohort\": pl.String})\n",
    "\n",
    "# Read autosomes\n",
    "with open(config[\"resources\"] + \"ref_fna/autosomes.list\") as f:\n",
    "    chromosomes = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge ROH files and filter\n",
    "\n",
    "# Read colony info\n",
    "#colonies_file = \"/master/abagwell/variant-analysis/resources/rhesus/pop/colonies.tsv\"\n",
    "colonies_file = config[\"cohorts\"]\n",
    "colonies = pl.read_csv(colonies_file, separator=\"\\t\", infer_schema_length=None, schema_overrides={\n",
    "    \"Dam\": pl.String,\n",
    "    \"Sire\": pl.String,\n",
    "    \"Id\": pl.String,\n",
    "    \"Cohort\": pl.String,\n",
    "})\n",
    "\n",
    "# Read chromosome lengths\n",
    "chromosome_df = pl.read_csv(f\"{config['resources']}ref_fna/chromosomes.tsv\", comment_prefix='#', separator='\\t')\n",
    "chromosome_lengths = list(chromosome_df[\"Length\"])\n",
    "chromosomes = chromosome_df[\"Chromosome\"]\n",
    "genome_length = sum(chromosome_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## For when using `bcftools roh`\n",
    "# Changes column names to match scikit-allel\n",
    "\n",
    "# Variables to adjust\n",
    "min_ROH_length = 1_000_000\n",
    "#min_ROH_length = 0\n",
    "\n",
    "\n",
    "roh_dfs = []\n",
    "for chrom in chromosomes:\n",
    "    roh_file = f\"{config['results']}roh/bcftools/{config['dataset']}.all.chr{chrom}.RG.roh\"\n",
    "    roh_df = pl.read_csv(roh_file, comment_prefix=\"$\", separator=\"\\t\", skip_lines=3,\n",
    "        schema_overrides=[pl.String, pl.String, pl.String, pl.Int32, pl.Int32, pl.Int64, pl.Int32, pl.Float32],\n",
    "        new_columns=[\"RG\", \"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Number of markers\", \"Quality (average fwd-bwd phred score)\"]\n",
    "    ).filter(\n",
    "        # Keep only large ROHs\n",
    "        pl.col(\"length\") > min_ROH_length\n",
    "    ).with_columns(\n",
    "    )#.join(colonies, how=\"left\", left_on=\"sample\", right_on=\"Id\")\n",
    "    \n",
    "    roh_dfs.append(roh_df)\n",
    "\n",
    "roh_df = pl.concat(roh_dfs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data per animal for source groups\n",
    "per_indiv_df = roh_df.select(\"sample\", \"chrom\", \"start\", \"stop\", \"length\"\n",
    "# .filter(\n",
    "#     # Use a filter if we wanting to keep only certain cohorts\n",
    "#     ~pl.col(\"Cohort\").is_in([\"Conventional source\", \"Brooks source\", \"NEPRC source\"])\n",
    ").group_by(\"sample\").agg(\n",
    "    froh = pl.sum(\"length\").truediv(genome_length),\n",
    "    number_of_ROHs = pl.count(\"length\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: It's very important that the animals listed in `colonies` actually were run through the fROH analysis;\n",
    "# otherwise, they will show up as having 0 ROHs when they really might still have some.\n",
    "all_indiv_df = colonies.join(per_indiv_df, how=\"left\", left_on=\"Id\", right_on=\"sample\").with_columns(\n",
    "# Index populations to be in correct order (TODO: Does this actually do anything?)\n",
    "    pl.col(\"Cohort\").cast(pl.Enum(\n",
    "        colors[\"Cohort\"]\n",
    "    )),\n",
    "    # Assign 0 to the samples not found to have ROHs\n",
    "    pl.col(\"froh\").fill_null(0),\n",
    "    pl.col(\"number_of_ROHs\").fill_null(0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.Chart(all_indiv_df).mark_circle(opacity=0.9).encode(\n",
    "    alt.X(\"number_of_ROHs\", title=\"Number of ROHs\"),\n",
    "    alt.Y(\"froh\", title=\"fROH\"),\n",
    "    alt.Color(\"Cohort\", title=\"Cohorts\").scale(\n",
    "        domain = list(colors[\"Cohort\"]),\n",
    "        range = list(colors[\"Color\"])\n",
    "    )\n",
    ").properties(\n",
    "    title=\"Individuals by ROH\"\n",
    ")#.save(\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.by_indiv.min1Mb.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# roh_df = roh_df.drop_nulls(\"Cohort\"\n",
    "roh_df = all_indiv_df.drop_nulls(\"Cohort\"\n",
    ").group_by(\"Cohort\").agg(\"*\").with_columns(\n",
    "# Index populations so that year ranges have their own color\n",
    "    pl.col(\"Cohort\").cast(pl.Enum(\n",
    "        colors[\"Cohort\"]\n",
    "    )),\n",
    ").with_row_index(\"pop_idx\", offset=1).with_columns(\n",
    "    # Find which are year ranges\n",
    "    is_year = pl.col(\"Cohort\").cast(pl.String).str.contains(\"-\").not_().cast(pl.Int8)\n",
    ").with_columns(\n",
    "    # Set the index of year ranges to 0\n",
    "    pl.col(\"pop_idx\").mul(\"is_year\")\n",
    ").drop(\"is_year\").sort(\"Cohort\"#.explode(pl.exclude(\"pop_idx\"))\n",
    ").group_by(\n",
    "    # Create color index\n",
    "    \"pop_idx\", maintain_order=True\n",
    ").agg('*').with_row_index(\"color_idx\").drop(\"pop_idx\").explode(pl.exclude(\"color_idx\")).explode(pl.exclude(\"Cohort\", \"color_idx\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Plot ROH lengths by group\n",
    "\n",
    "# # mean_roh_df = roh_df.group_by(\"Cohort\", \"color_idx\").agg(pl.mean(\"length\"))\n",
    "# mean_roh_df = roh_df.group_by(\"Cohort\").agg(pl.mean(\"length\"))\n",
    "\n",
    "# # # Index populations so that year ranges have their own color\n",
    "# # ).with_columns(\n",
    "# #     pl.col(\"Cohort\").cast(pl.Enum(\n",
    "# #         colors[\"Cohort\"]\n",
    "# #     )),\n",
    "# # ).with_row_index(\"pop_idx\", offset=1).with_columns(\n",
    "# #     # Find which are year ranges\n",
    "# #     is_year = pl.col(\"Cohort\").cast(pl.String).str.contains(\"-\").not_().cast(pl.Int8)\n",
    "# # ).with_columns(\n",
    "# #     # Set the index of year ranges to 0\n",
    "# #     pl.col(\"pop_idx\").mul(\"is_year\")\n",
    "# # ).drop(\"is_year\").sort(\"Cohort\").group_by(\n",
    "# #     # Create color index\n",
    "# #     \"pop_idx\", maintain_order=True\n",
    "# # ).agg('*').with_row_index(\"color_idx\").drop(\"pop_idx\"\n",
    "# # ).explode(\"Cohort\", \"length\")\n",
    "\n",
    "# mean_roh_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# alt.Chart(mean_roh_df).mark_bar().encode(\n",
    "#     alt.X(\"Cohort\", title=\"Cohort\"),\n",
    "#     alt.Y(\"length\", title=\"Mean Length\"),\n",
    "#     alt.Color(\"Cohort:N\", title=None, legend=None).scale(\n",
    "#         domain = list(colors[\"Cohort\"]),\n",
    "#         range = list(colors[\"Color\"])\n",
    "#     ),\n",
    "# ).properties(\n",
    "#     title=\"ROH Across Cohorts\"\n",
    "# )#.save(\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.mean_roh.min1Mb.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_trio_roh(df):\n",
    "    \"\"\"Plot the ROHs across all chromosomes of one trio.\"\"\"\n",
    "\n",
    "    # Create dummy records. That is, an ROH of length 0 for every combination of sample and chromosome so that they all appear on the plot\n",
    "    chromosomes = roh_df.select(\"chrom\").unique()\n",
    "    sequenced_animals = df.select(\"sample\", \"Colony\").unique()\n",
    "    dummy_records = sequenced_animals.join(chromosomes, how=\"cross\").with_columns(\n",
    "        start = pl.lit(0).cast(pl.Int32),\n",
    "        stop = pl.lit(0).cast(pl.Int32),\n",
    "        length = pl.lit(0).cast(pl.Int64),\n",
    "    ).select(  # Reorder\n",
    "        \"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Colony\"\n",
    "    )\n",
    "\n",
    "    # Concatenate ROHs with dummy records\n",
    "    df = pl.concat([df, dummy_records]).sort(\"sample\")  # TODO: Improve sorting for when needs to be numeric or animal ID has \"X\" in it\n",
    "\n",
    "    columns = []\n",
    "    for idx, chrom in enumerate(chromosomes):\n",
    "        #column_plots = []\n",
    "        # Only keep y-axis labels on leftmost plot\n",
    "        if idx == 0:\n",
    "            axis = alt.Axis()\n",
    "        else:\n",
    "            axis = None\n",
    "        # # Info from `bcftools roh`\n",
    "        chrom_df = df.filter(\n",
    "            # Keep only large ROHs\n",
    "            #pl.col(\"length\") > 2_000_000\n",
    "            pl.col(\"chrom\") == chrom\n",
    "        )\n",
    "\n",
    "        roh_plot = alt.Chart(chrom_df).mark_bar().encode(\n",
    "        alt.X(\"mb_start_pos:Q\", title=[\"Position\", \"(Mb)\"]).scale(domainMin=0, domainMax=chromosome_lengths[int(chrom) -1]/1_000_000, clamp=True),\n",
    "        alt.X2(\"mb_stop_pos:Q\"),\n",
    "        alt.Y(\"sample:O\", axis=axis, title=\"Sample\", sort=alt.EncodingSortField(field='animal', order='ascending')), # sort=alt.EncodingSortField(field='date', order='ascending')\n",
    "        color = alt.Color(\"Colony:N\").scale(\n",
    "            domain = list(colors[\"Cohort\"]),\n",
    "            range = list(colors[\"Color\"])\n",
    "        ),\n",
    "        #order=alt.Order('animal:O', sort='ascending'),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(\"chrom\", title=\"Chr\"),\n",
    "            alt.Tooltip(\"start\", title=\"Start (bp)\"),\n",
    "            alt.Tooltip(\"stop\", title=\"End (bp)\"),\n",
    "            alt.Tooltip(\"length\", title=\"Length\"),\n",
    "        ]\n",
    "        ).properties(\n",
    "            width=chromosome_lengths[int(chrom) - 1]/1500000,\n",
    "            height=alt.Step(10),\n",
    "            #width=300, # For when only displaying one chromosome\n",
    "            #title=[\"RoH\", f\"chr{chrom}\"]\n",
    "        ).transform_calculate(\n",
    "            mb_start_pos = 'datum.start / 1000000',\n",
    "            mb_stop_pos = 'datum.stop / 1000000',\n",
    "        )\n",
    "    \n",
    "        #plot_list.append(roh_plot)\n",
    "        columns.append(roh_plot)\n",
    "    \n",
    "    return alt.hconcat(*columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_all_trio_roh(df):\n",
    "    # Find sequenced trios\n",
    "    sequenced_animals = df[\"sample\"].unique()\n",
    "    trios = colonies.filter(\n",
    "        pl.col(\"Id\").is_in(sequenced_animals) & pl.col(\"Sire\").is_in(sequenced_animals) & pl.col(\"Dam\").is_in(sequenced_animals)\n",
    "    ).select(\"Id\", \"Sire\", \"Dam\").sort(\"Id\")\n",
    "\n",
    "    # Pull out specific trio\n",
    "    rows = []\n",
    "    for trio in trios.transpose().select([f\"column_{num}\" for num in range(50,75)]):\n",
    "        trio_df = roh_df.filter(\n",
    "            pl.col(\"sample\").is_in(trio)\n",
    "        ).select(\"sample\", \"chrom\", \"start\", \"stop\", \"length\", \"Colony\")\n",
    "        \n",
    "        rows.append(plot_trio_roh(trio_df))\n",
    "    return alt.vconcat(*rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dataframe for subsequent plots\n",
    "\n",
    "chrom_df = pl.DataFrame({\n",
    "    \"chrom\": chromosomes,\n",
    "    \"chrom_len\": chromosome_lengths\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # total_len_roh = roh_df.group_by(\"sample\", \"color_idx\", \"Cohort\").agg(pl.sum(\"length\")).sort(\"sample\").with_columns(\n",
    "# total_len_roh = roh_df.sort(\"Id\").with_columns(\n",
    "#     pl.col(\"length\").truediv(genome_length).alias(\"froh\")\n",
    "# # ).with_columns(\n",
    "# #     # Add seq info for splitting by seq\n",
    "# #     seq = pl.col(\"sample\").str.slice(0, 3)\n",
    "# # )\n",
    "# # .with_columns(\n",
    "# #     pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3).alias(\"animal\")\n",
    "# ).group_by(\"sample\").agg(pl.last(\"froh\"), pl.last(\"color_idx\"), pl.last(\"Cohort\"))#.with_columns(\n",
    "#     # Pull out the WGS over WES and select first\n",
    "#     #pl.col(\"sample\").list.sort(descending=True).list.first(),\n",
    "#     #pl.col(\"froh\").list.sort(descending=True).list.first(),\n",
    "# # ).with_columns(\n",
    "# #     cohort = pl.concat_str([pl.col(\"Colony\"), pl.col(\"Cohort\")], separator=\"_\")\n",
    "#     #pl.col(\"sample\").str.split(\"_\").list.get(0).str.slice(3),\n",
    "# # ).filter(\n",
    "# #     ~pl.col(\"Cohort\").is_null()\n",
    "# # )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boxplot of fROH over Cohorts\n",
    "boxplot = alt.Chart(roh_df).mark_boxplot().encode(\n",
    "    alt.X(\"Cohort:N\", title=\"Cohort\", #axis=alt.Axis(labelAngle=-45),\n",
    "        sort=colors[\"Cohort\"],   \n",
    "    ),\n",
    "    alt.Y(\"froh:Q\", title=\"fROH\"),\n",
    "    #alt.Column(\"seq\", title=\"Sequence Type\"),\n",
    "    color=alt.Color(\"Cohort:N\", legend=None).scale(\n",
    "        domain = list(colors[\"Cohort\"]),\n",
    "        range = list(colors[\"Color\"])\n",
    "    )\n",
    ").properties(\n",
    "    #width=100\n",
    "    title=\"fROH in Cohorts\"\n",
    ")\n",
    "\n",
    "boxplot#.save(\"/master/abagwell/figures/final_plots/U42_WES.common_between_founding_cohorts2.froh_boxplot.min1Mb.html\")\n",
    "\n",
    "#boxplot.save(f\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.boxplot.maf_0.05.gt{min_ROH_length / 1_000_000}Mb.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Violin plot (without error bars or means). See next code block for it with extra details\n",
    "\n",
    "violin = alt.Chart(roh_df\n",
    "#                    .with_columns(\n",
    "#    pl.col(\"Cohort\").str.replace(\"Offspring of merger\", \"Merged\"))\n",
    ").transform_density(\n",
    "    'froh',\n",
    "    as_=['froh', 'density'],\n",
    "    #extent=[0, 0.05],\n",
    "    #extent=[0, 0.1],\n",
    "    groupby=['Cohort', 'color_idx']\n",
    ").mark_area(orient='horizontal').encode(\n",
    "    alt.X(\"density:Q\").stack('center')\n",
    "        .impute(None)\n",
    "        .title(None)\n",
    "        .axis(labels=False, values=[0], grid=False, ticks=True),\n",
    "    alt.Y(\"froh:Q\", title=\"fROH\"),\n",
    "    alt.Column(\"Cohort:N\", title=\"Cohort\",\n",
    "          # TODO: Generalize this\n",
    "        sort=colors[\"Cohort\"],   \n",
    "        ).spacing(0).header(titleOrient='bottom', labelOrient='bottom', labelPadding=0),\n",
    "    color=alt.Color(\"Cohort:N\").scale(\n",
    "        domain = list(colors[\"Cohort\"]),\n",
    "        range = list(colors[\"Color\"])\n",
    "    ),\n",
    ").properties(\n",
    "    width=65,\n",
    "    title=\"fROH Across Cohorts\"\n",
    ")\n",
    "\n",
    "\n",
    "violin\n",
    "\n",
    "#violin.save(f\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.violinplot.maf_0.05.gt{min_ROH_length / 1_000_000}Mb.full_height.html\")\n",
    "#violin.save(f\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.violinplot.maf_0.05.gt{min_ROH_length}Mb.max_0.1.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Violin plot but with error bars and means\n",
    "# Has to be more complicated in order to construct a layered chart that also is faceted\n",
    "# Variations on the graph can be made my adjusting the scale on the alt.Y of the violin plot\n",
    "\n",
    "# Varibles to adjust\n",
    "error_unit = 'stderr' # Can switch extent to `stdev`, `stderr`, or `ci`\n",
    "# max_y = 0.08\n",
    "# min_y = -0.02\n",
    "#max_y = 0.055\n",
    "#max_y = 0.09\n",
    "max_y = 0.6\n",
    "#max_y=0.14 # For zoomed of Merger\n",
    "min_y = 0.0\n",
    "\n",
    "violin = alt.Chart().transform_density(\n",
    "    'froh',\n",
    "    as_=['froh', 'density'],\n",
    "    extent=[min_y, max_y],\n",
    "    groupby=['Cohort', 'color_idx']\n",
    ").mark_area(orient='horizontal').encode(\n",
    "    alt.X(\"density:Q\").stack('center')\n",
    "        .impute(None)\n",
    "        .title(None)\n",
    "        .axis(labels=False, values=[0], grid=False, ticks=True)\n",
    "        .scale(nice=False,zero=False),\n",
    "    #alt.X(\"Cohort\"),\n",
    "    #alt.Y(\"froh:Q\", title=x_title).axis(range=[0, 0.05]),\n",
    "    alt.Y(\"froh:Q\", title='fROH'),#.scale(domain=[min_y, max_y]),\n",
    "    # alt.Column(\"Cohort:N\", title=\"Cohort\",\n",
    "    #       # TODO: Generalize this\n",
    "    #     ).spacing(0).header(titleOrient='bottom', labelOrient='bottom', labelPadding=0),\n",
    "    color=alt.Color(\"Cohort:N\", legend=None).scale(\n",
    "        domain = list(colors[\"Cohort\"]),\n",
    "        range = list(colors[\"Color\"])\n",
    "    )\n",
    ").properties(\n",
    "    width=25\n",
    ")\n",
    "\n",
    "error = alt.Chart().mark_errorbar(extent=error_unit).encode(\n",
    "    #alt.X('Cohort', title=None),\n",
    "    alt.Y('froh:Q', title='fROH')\n",
    "    )\n",
    "\n",
    "mean = alt.Chart().mark_circle(color='black').encode(\n",
    "    #alt.X('Cohort', title=None),\n",
    "    alt.Y('mean(froh):Q', title='fROH')\n",
    "    )\n",
    "\n",
    "layered = alt.layer(violin, error, mean, data=roh_df\n",
    "        # .filter(\n",
    "        #     pl.col(\"Cohort\").is_in([\"2018-2020\", \"Offspring of merger\", \"NEPRC source\"]))\n",
    "        #     #pl.col(\"Cohort\").is_in([\"Conventional source\", \"Brooks source\", \"NEPRC source\"]))\n",
    ").facet(\n",
    "    #column='Cohort'\n",
    "    alt.Column('Cohort',\n",
    "        header=alt.Header(\n",
    "            labelOrient='bottom', labelPadding=0, labelAnchor='middle', labelAngle=-90, labelBaseline=\"middle\", labelAlign=\"right\",\n",
    "            title='Cohort', titleAlign=\"center\", titleOrient='bottom', ) # labelAngle=-45\n",
    "    )\n",
    ").resolve_scale(x=alt.ResolveMode(\"independent\")\n",
    ").configure_facet(\n",
    "    spacing=0,\n",
    ").configure_title(anchor='middle').properties(\n",
    "    title=\"fROH in Cohorts\"\n",
    "    #title=\"fROH of Merger\"\n",
    "    #title=[\"fROH in\", \"Founding Populations\"]\n",
    ")\n",
    "\n",
    "\n",
    "layered#.save(\"/master/abagwell/figures/final_plots/narrow_violins/U42_WES.common_between_founding_cohorts2.violinplot_froh.min1Mb.merger_zoomed.html\")\n",
    "layered#.save(f\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.froh_violinplot.gt{min_ROH_length / 1_000_000}Mb.{error_unit}.full_height.html\")\n",
    "#layered.save(f\"/master/abagwell/figures/roh/U42_WES.common_between_founding_cohorts2.froh_violinplot.{error_unit}.html\")\n",
    "#layered.save(f\"{config[\"results\"]}figures/{config[\"dataset\"]}.all.froh_violinplot.{error_unit}.cutoff.html\")\n",
    "#layered.save(f\"{config[\"results\"]}figures/{config[\"dataset\"]}.all.froh_violinplot.{error_unit}.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "\n",
    "import pandas as pd\n",
    "import scipy\n",
    "\n",
    "# Do T-test for each pair of cohorts\n",
    "results = []\n",
    "for cohort1, cohort2 in combinations(list(roh_df[\"Cohort\"].unique()), 2):\n",
    "    result = scipy.stats.ttest_ind(\n",
    "        roh_df.filter(\n",
    "            pl.col('Cohort') == cohort1\n",
    "    )['froh'],\n",
    "        roh_df.filter(\n",
    "            pl.col('Cohort') == cohort2\n",
    "    )['froh'],\n",
    "    )\n",
    "    results.append((cohort1, cohort2, result.statistic, result.pvalue, result.df))\n",
    "\n",
    "df = pd.DataFrame(results, columns=['Cohort1', 'Cohort2', 'Statistc', 'p-Value', 'DF'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graph2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
